---
title: "DNA methylation analysis code_R"
author: "Dantong Zhu"
date: "2026-01-06"
output:
  html_document:
    css: style.css
    fig_caption: yes
    highlight: tango
    keep_md: yes
    toc: yes
    toc_float:
      collapsed: no
    toc_depth: 3
    self_contained: yes
    code_folding: hide
  word_document:
    toc: yes
    toc_depth: '3'
  pdf_document:
    toc: yes
    toc_depth: '3'
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE,
                      eval = TRUE)

library(ggplot2)
library(dplyr)
library(ggfortify)
library(ggpubr)
library(pheatmap)
library(tidyr)
library(readxl)
library(tidyverse)
library(stringr)
library(caret)
library(edgeR)
library(limma)
library(ComplexUpset)
library(ComplexHeatmap)
library(org.Mm.eg.db)
library(splines)
library(ggnewscale)
library(DMRcate)
library(plyr)
library(rGREAT)
library(msigdbr)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(glmnet)
library(ggrastr)
library(reshape2)
library(khroma)
library(knowYourCG)
library(SummarizedExperiment)
library(sesameData)
library(gprofiler2)
library(WGCNA)
library(lme4)
library(lmerTest)
library(ggforce)
library(ppcor)
library(lmtest)
library(variancePartition)
library(ggVennDiagram)
```

# 1. Import data; global methylation pattern and pca
```{r}
load("~/Desktop/DNA methylation/discovery_cohort_DNAm_data.RData") #methylation matrix data
load("~/Desktop/DNA methylation/validation_cohort_DNAm_data.RData")
load("~/Desktop/DNA methylation/methy_prob_loci.RData") # prob loci
source("~/Desktop/DNA methylation/DNA_methylation_functions.R") # functions for this analysis
beta2m = function(x) {log2(x/(1-x))}
methy_dat_dis_M = methy_dat_dis %>% apply(., 2, beta2m) %>% as.data.frame() #convert to Mvalues
methy_dat_val_M = methy_dat_val %>% apply(., 2, beta2m) %>% as.data.frame()
```

## 1.1. Frailty index visualization
```{r}
ggplot2::theme_set(theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.title = element_text(hjust = 0.5), text = element_text(size = 8), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1.2),
        axis.text.y = element_text(color = "black"),
        axis.text.x = element_text(color = "black")))
methy_dat_cg_pca = prcomp(methy_dat_dis_M, scale. = TRUE) # pca
ggplot(data = methy_dat_dis_meta,
       aes(x = Time.label,
           y = FI.Score,
           color = Sex)) +
   geom_boxplot() +
   scale_color_manual(values = c("#AA4466", "#4477AA")) ## FI boxplot
```

```{r}
ggplot(data = methy_dat_val_meta,
       aes(x = Time.label,
           y = FI.Score,
           color = Sex)) + 
  geom_boxplot() +
   scale_color_manual(values = c("#AA4466", "#4477AA")) ## FI boxplot validation
```

## 1.2. Distribution of beta values
```{r}
methy_dat_cg_melt = methy_dat_dis %>% 
                    mutate(id = rownames(.),
                           sex = methy_dat_dis_meta$Sex,
                           tp = methy_dat_dis_meta$Time.label) %>% 
                    melt(., id = c("id", "sex", "tp")) %>% as.data.frame() %>%
                    mutate(group = paste(sex, tp))

ggplot(data = methy_dat_cg_melt, # density plot
       aes(x = value,
           color = tp,
           linetype = sex)) +
    geom_density(linewidth = 1) +
    scale_color_manual(values = c("#332288", "#AAAA00", "#44AA99", "#882255", "#88CCEE"))
```

```{r}
methy_dat_cg_melt_med = methy_dat_cg_melt %>%
                  dplyr::group_by(group) %>%
                  dplyr::mutate(med = median(value)) %>%
                  as.data.frame() %>%
                  dplyr::select(group, med, sex) %>%
                  unique()

ggplot(data = methy_dat_cg_melt, #violin plot
       aes(x = group,
           y = value,
           color = group)) +
    geom_violin(aes(linetype = group)) +
    geom_boxplot(width = 0.1) +
    scale_color_manual(values = c("#332288", "#AAAA00", "#44AA99", "#882255",  
                                  "#332288", "#AAAA00", "#44AA99", "#882255", "#88CCEE")) +
    scale_linetype_manual(values = c(1, 1, 1, 1,
                             2, 2, 2, 2, 2)) +
    geom_point(data = methy_dat_cg_melt_med,
               aes(x = group, y = med)) +
    geom_line(data = methy_dat_cg_melt_med,
              aes(x = group, y = med, group = sex), color = "grey30", lty = 2) + 
    labs(x = "")
```

## 1.3. PCA analysis
```{r}
## PCs by sex and time points
ggplot(data = methy_dat_cg_pca$x,
       aes(x = PC1,
           y = PC2,
           color = methy_dat_dis_meta$Time.label,
           shape = methy_dat_dis_meta$Sex)) +
   geom_point(size = 3) + 
   scale_color_manual(values = c("#332288", "#AAAA00", "#44AA99", "#882255", "#88CCEE")) +
   labs(x = "PC1", y = "PC2", 
        color = "Time points", 
        shape = "Sex") 
```

```{r}
## PCs by logFI
ggplot(data = methy_dat_cg_pca$x,
       aes(x = PC1,
           y = PC2,
           color = log2(methy_dat_dis_meta$FI.Score),
           shape = methy_dat_dis_meta$Sex)) +
   geom_point(size = 3) + 
   scale_color_bamako(midpoint = -2.5)  +
   labs(x = "PC1", y = "PC2", 
        color = "log2(Frailty Index)", 
        shape = "Sex") 
```

## 1.4. Variance partitioning analysis 
```{r}
pc_matrix = methy_dat_cg_pca$x[, 1:121] %>% t() %>% as.data.frame()
methy_dat_dis_meta = methy_dat_dis_meta %>%
  mutate(logFI = log2(FI.Score)) %>%
  mutate(fi_resid = residuals(lm(logFI ~ `Age.at.assesment..days.`, data = methy_dat_dis_meta)),
         mouse_id = as.factor(Mouse.ID),
         sex = ifelse(Sex == "female", 1, 0))
var_partition_res = fitExtractVarPartModel(pc_matrix, 
                                           ~ fi_resid + `Age.at.assesment..days.` + sex + (1 |mouse_id),
                                           methy_dat_dis_meta)
var_partition_visual_tbl = rbind(c(sum(summary(methy_dat_cg_pca)$importance[2, 1:121]*var_partition_res[, 1]),
                                   sum(summary(methy_dat_cg_pca)$importance[2, 1:121]*var_partition_res[, 2]),
                                   sum(summary(methy_dat_cg_pca)$importance[2, 1:121]*var_partition_res[, 3]),
                                   sum(summary(methy_dat_cg_pca)$importance[2, 1:121]*var_partition_res[, 4])),
                                 var_partition_res[1:15, 1:4]) %>% as.data.frame() %>%
  dplyr::select(mouse_id, sex, `Age.at.assesment..days.`, fi_resid)

pheatmap((var_partition_visual_tbl)*100 %>% as.matrix(),
         border_color = "white",
         colorRampPalette(c("grey90", "#6699CC", "#332288"))(30),
         cluster_cols = F,
         cluster_rows = F,
         angle_col = "45",
         show_rownames = TRUE)
```

# 2. EWAS analysis on sex-inclusive subgroups

## 2.1. limma analysis with random effect to get DMPs
```{r}
## limma analysis 
metadata_fm_excl = methy_dat_dis_meta %>% filter(Time.label != "T5")
sex = factor(metadata_fm_excl$Sex)
frail = log2(metadata_fm_excl$FI.Score)
design_fm_frail = model.matrix(~ 0 + sex + sex:frail)
colnames(design_fm_frail) = make.names(colnames(design_fm_frail))
methy_dat_fm_excl = methy_dat_dis_M %>% filter(rownames(.) %in% metadata_fm_excl$id)
methy_counts_excl = methy_dat_fm_excl %>% t() %>% as.data.frame() 
corfit = duplicateCorrelation(methy_counts_excl, design_fm_frail, block = metadata_fm_excl$Mouse.ID)
fm_frail_fit2_res = get_fit2_res(metadata_fm_excl, design_fm_frail, methy_dat_dis_M)
mix_contrast = makeContrasts((sexfemale.frail + sexmale.frail)/2,
                         levels = colnames(coef(fm_frail_fit2_res)))
mix_tt = contrasts.fit(fm_frail_fit2_res, mix_contrast) %>% 
  eBayes() %>%
  topTable(., n = Inf, adjust.method = "BH", sort.by = "none")
mix_tt_dmp = mix_tt %>% filter(adj.P.Val < 0.05)
```

```{r}
## manhattan plot for DMPs
cpg_pos = merge(mix_tt %>% mutate(IlmnID = rownames(.)), cpg_loci, by = "IlmnID") %>% 
    as.data.frame() %>%
    dplyr::select(IlmnID, chromo, MAPINFO, P.Value, adj.P.Val)
cpg_pos_cumpos = merge(cpg_pos, chromo_cumpos, by = "chromo") %>%
    mutate(bpcum = MAPINFO + position_x) %>%
    dplyr::select(chromo, IlmnID, P.Value, adj.P.Val, MAPINFO, position_x, bpcum)
cpg_pos_cumpos$chromo = factor(cpg_pos_cumpos$chromo, levels = chrOrder_DNAm)
nondmp_pos_cumpos = cpg_pos_cumpos %>% filter(adj.P.Val >= 0.05)
dmp_pos_cumpos = cpg_pos_cumpos %>% filter(adj.P.Val < 0.05)
x_axis_ticks = cpg_pos_cumpos %>%
    dplyr::group_by(chromo) %>%
    dplyr::summarise(center = (max(bpcum) + min(bpcum))/2)
mix_manhattan_plot_loci = ggplot() +
    geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = 0, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
    geom_point(data = nondmp_pos_cumpos,
               aes(x = bpcum,
                   y = -log10(P.Value),
                   color = as.factor(chromo)), size = 0.2) + 
    scale_color_manual(values = rep(c("grey65", "grey80"), 11)) +
    guides(color = FALSE) + 
    new_scale_color() + 
    geom_point(data = dmp_pos_cumpos, 
               aes(x = bpcum,
                   y = -log10(P.Value), 
                   color = as.factor(chromo)), size = 0.4) + 
    scale_color_manual(values = rep(c("#114477", "#44AA99"), 11)) + 
    scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
    labs(x = "", y = "-log10(p-value)", color = "") +
    guides(color = FALSE) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 12, face = "bold"))
rasterize(mix_manhattan_plot_loci, dpi = 300)
dev.off()
```

## 2.2. Proportions in autosome and sex chromosomes (chi-squared test)
```{r}
## comparisons of cpg and gene location categories
mix_tt_cpg_loci = merge(mix_tt %>% mutate(IlmnID = rownames(.)),
                        cpg_loci,
                        by = "IlmnID")
mix_tt_cpg_loci_chrx = mix_tt_cpg_loci %>% filter(chromo == "chrX") # n = 10338
mix_tt_cpg_loci_chrx_dmp = mix_tt_cpg_loci_chrx %>% filter(adj.P.Val < 0.05) # n = 2184

mix_tt_cpg_loci_chry = mix_tt_cpg_loci %>% filter(chromo == "chrY") # n = 94
mix_tt_cpg_loci_chry_dmp = mix_tt_cpg_loci_chry %>% filter(adj.P.Val < 0.05) # n = 10

mix_tt_cpg_loci_chrmt = mix_tt_cpg_loci %>% filter(chromo == "chrMT") # n = 35
mix_tt_cpg_loci_chrmt_dmp = mix_tt_cpg_loci_chrmt %>% filter(adj.P.Val < 0.05) # n = 4

mix_tt_cpg_loci_autosome = mix_tt_cpg_loci %>% filter(!chromo %in% c("chrX", "chrY", "chrMT")) 
mix_tt_cpg_loci_autosome_dmp = mix_tt_cpg_loci_autosome %>% filter(adj.P.Val < 0.05)
```

```{r}
load("~/cpg_annotation_tbl.RData") 
cpg_dat = read.csv("~/MouseMethylation-12v1-0_A2.csv") # downloaded from Illumina web
colnames(cpg_annotation_tbl) = c("Name", "CpG_context", "Gene_context")
cpg_annotation_tbl = cpg_annotation_tbl %>%
  mutate(Gene_context_u = case_when(
    Gene_context == "Shore" ~ "tss_1500",
    TRUE ~ Gene_context))

cpg_dat_context_anno = merge(cpg_dat, cpg_annotation_tbl, by = "Name", all.x = TRUE) %>%
  as.data.frame() %>%
  filter(Probe_Type == "cg") %>%
  dplyr::select(IlmnID, CpG_context, Gene_context_u) %>%
  mutate(Gene_context = Gene_context_u) %>%
  dplyr::select(-Gene_context_u)

mix_all_anno = cpg_dat_context_anno %>%
  filter(IlmnID %in% mix_tt_cpg_loci$IlmnID)
mix_dmp_all_anno = cpg_dat_context_anno %>%
  filter(IlmnID %in% rownames(mix_tt_dmp))
mix_dmp_autosome_anno = cpg_dat_context_anno %>% 
  filter(IlmnID %in% mix_tt_cpg_loci_autosome_dmp$IlmnID)
mix_dmp_chrx_anno = cpg_dat_context_anno %>% 
  filter(IlmnID %in% mix_tt_cpg_loci_chrx_dmp$IlmnID)

mix_context_anno = Reduce(rbind, list(mix_all_anno %>% mutate(type = "all"),
                                      mix_dmp_all_anno %>% mutate(type = "dmp"),
                                      mix_dmp_autosome_anno %>% mutate(type = "autosome"),
                                      mix_dmp_chrx_anno %>% mutate(type = "chrx"))) %>%
                                      as.data.frame() 

mix_context_anno_cpg_dmp = mix_context_anno %>%
  dplyr::group_by(type, CpG_context) %>%
  dplyr::tally() %>%
  dplyr::mutate(prop_cpg = n/sum(n))

mix_context_anno_gene_dmp = mix_context_anno %>%
  dplyr::group_by(type, Gene_context) %>%
  dplyr::tally() %>%
  dplyr::mutate(prop_gene = n/sum(n))

mix_context_anno_cpg_dmp$CpG_context = factor(mix_context_anno_cpg_dmp$CpG_context, 
                                          levels = c("Island", "Shore", "Shelf", "Open Sea"))
mix_context_anno_cpg_dmp$type = factor(mix_context_anno_cpg_dmp$type,
                                   levels = c("all", "dmp", "autosome", "chrx"))
mix_cpg_context_plot = ggplot(data = mix_context_anno_cpg_dmp,
       aes(x = type,
           y = prop_cpg,
           fill = CpG_context)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_fill_manual(values = c("#DDCC77", "#CC6677", "#AA4499", "#882255"))

mix_context_anno_gene_dmp$Gene_context = factor(mix_context_anno_gene_dmp$Gene_context, 
                                              levels = c("tss_200", "tss_1500", "tss_body", "Intergenic"))
mix_context_anno_gene_dmp$type = factor(mix_context_anno_gene_dmp$type,
                                   levels = c("all", "dmp", "autosome", "chrx"))
mix_gene_context_plot = ggplot(data = mix_context_anno_gene_dmp,
       aes(x = type,
           y = prop_gene,
           fill = Gene_context)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_fill_manual(values = c("#DDCC77", "#CC6677", "#AA4499", "#882255"))

ggarrange(mix_cpg_context_plot, mix_gene_context_plot,
          nrow = 1, ncol = 2) ## content boxplot
```

## 2.3. kernal smoothing to get DMRs
```{r}
## plot dmrs_mix
mix_cpg_loci = methy_prob_loci %>%
    filter(IlmnID %in% colnames(methy_dat_dis_M)) %>%
    mutate(chromo = paste0("chr", CHR))
mix_cpg_fc = merge(mix_tt %>% mutate(IlmnID = rownames(.)), mix_cpg_loci, by = "IlmnID")
mix_dmr_all = get_dmr_modify_DMRcate(mix_cpg_fc) 
mix_dmr_greater5 = mix_dmr_all %>% filter(no.cpgs >= 5)

mix_dmr_stouffer = mix_dmr_greater5 %>%
  dplyr::select(coord, no.cpgs, chr, start, end, Stouffer) %>%
  as.data.frame() %>%
  mutate(MAPINFO = (as.numeric(start) + as.numeric(end))/2) %>%
  mutate(chromo = chr) %>%
  merge(., chromo_cumpos, by = "chromo") %>%
  mutate(bpcum = MAPINFO + position_x) 

ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = 0, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = mix_dmr_stouffer %>% filter(Stouffer < 0.05),
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 size = log2(no.cpgs),
                 color = as.factor(chromo))) +
  scale_size(range = c(1, 3)) +
  scale_color_manual(values = rep(c("#114477", "#44AA99"), 11)) +
  new_scale_color() +
  geom_point(data = mix_dmr_stouffer %>% filter(Stouffer >= 0.05),
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 color = as.factor(chromo)),
             size = 0.2) +
  scale_color_manual(values = rep(c("grey65", "grey80"), 11)) +
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
  labs(x = "", y = "-log10(Stouffer)", color = "") + guides(color = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 12, face = "bold")) 
```

```{r}
## select dmrs by Stouffer
mix_dmr = mix_dmr_all %>% filter(no.cpgs >= 5) %>%
  filter(Stouffer < 0.05)
#get_DMRs(methy_dat_dis_M, mix_tt) # n = 1882 
## get cpgs within dmrs
mix_loci_dmrs = select_dmp_for_dmr(mix_dmr) # n = 12543
```

## 2.4. coMethDMR analysis, DMR level selection
```{r}
mix_loci_dmr_expr = methy_dat_dis_M %>% 
  filter(rownames(.) %in% metadata_fm_excl$id) %>%
  dplyr::select(mix_loci_dmrs$IlmnID) %>%
  mutate(id = rownames(.)) %>%
  merge(., metadata_fm_excl, by = "id") %>%
                  as.data.frame()

mix_loci_dmr_residual = lapply(2:(length(mix_loci_dmrs$IlmnID) + 1), function(x) {
  lm_fit = lm(mix_loci_dmr_expr[, x] ~ as.factor(Sex), data = mix_loci_dmr_expr)
  resid = residuals(lm_fit) %>% unname()
  return(resid)
}) %>% 
  Reduce(cbind, .) %>%
  as.data.frame() %>%
  `colnames<-`(mix_loci_dmrs$IlmnID)
rownames(mix_loci_dmr_residual) = mix_loci_dmr_expr$id
```

```{r}
mix_test_rdrop = test_rdrop(mix_loci_dmrs, mix_loci_dmr_residual)
mix_rdrop_loci_remain_v1 = mix_test_rdrop %>% filter(rdrop > 0.4)
mix_rdrop_dmr_above5 = table(mix_rdrop_loci_remain_v1$dmr) %>% as.data.frame() %>% filter(Freq >=5)
mix_rdrop_loci_remain_v2 = mix_rdrop_loci_remain_v1 %>% filter(dmr %in% mix_rdrop_dmr_above5$Var1)
```

```{r, message = FALSE, warning = FALSE}
## association by lmm, allowing random effect of CpGs
mix_loci_dmr_residual_meta = merge(mix_loci_dmr_residual %>% mutate(id = rownames(.)),
                           metadata_fm_excl, by = "id") %>% as.data.frame() %>%
  mutate(age = `Age.at.assesment..days.`,
         log2FI = log2(FI.Score))
mix_dmr_assoc_res = lapply(unique(mix_rdrop_loci_remain_v2$dmr), function(x) {
  loci_lmer = mix_rdrop_loci_remain_v2 %>% filter(dmr == x)
  expr_merged = mix_loci_dmr_residual_meta %>%
    dplyr::select(loci_lmer$IlmnID, id, Mouse.ID, age, log2FI) %>%
    melt(., id = c("id", "Mouse.ID", "age", "log2FI")) %>%
    as.data.frame()
  res1 = lmer(value ~ log2FI + age + (1 | Mouse.ID) + (log2FI | variable), data = expr_merged)
  res2 = lmer(value ~ age + (1 | Mouse.ID) + (age | variable), data = expr_merged)
  return(c(summary(res1)$coefficients[2, 5], summary(res2)$coefficients[2, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame()
rownames(mix_dmr_assoc_res) = unique(mix_rdrop_loci_remain_v2$dmr)
colnames(mix_dmr_assoc_res) = c("p_frail", "p_age")
mix_dmr_assoc_res_adj = mix_dmr_assoc_res %>%
  mutate(adjp_frail = p.adjust(p_frail, method = "fdr"),
         adjp_age = p.adjust(p_age, method = "fdr")) %>%
  mutate(cate = case_when(
    adjp_frail < 0.05 & adjp_age < 0.05 ~ "dual",
    adjp_frail < 0.05  ~ "frail",
    adjp_age < 0.05 ~ "age",
    TRUE ~ "other")) %>%
  mutate(dmr = rownames(.))
```

```{r}
mix_dmr_assoc_res_adj$cate = factor(mix_dmr_assoc_res_adj$cate, levels = c("frail", "dual", "age", "other"))
ggplot(data = mix_dmr_assoc_res_adj,
       aes(x = -log10(adjp_frail),
           y = -log10(adjp_age),
           color = as.factor(cate),
           shape = as.factor(cate))) +
  geom_point(size = 1.5) +
  geom_hline(yintercept = -log10(0.05), lty = 3, color = "grey50") +
  geom_vline(xintercept = -log10(0.05), lty = 3, color = "grey50") +
  scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99", "#999933")) +
  scale_shape_manual(values = c(19, 17, 15, 18)) +
  expand_limits(x = 0, y = 0) +
  labs(x = "-log10(p)_frail", 
       y = "-log10(p)_age",
       color = "")
```

```{r}
mix_frail_dmr_stouffer = mix_dmr_greater5 %>%
  dplyr::select(coord, no.cpgs, chr, start, end, Stouffer) %>%
  merge(., mix_dmr_assoc_res_adj, by.x = "coord", by.y = "dmr") %>%
  as.data.frame() %>%
  filter(cate != "other") %>%
  mutate(MAPINFO = (as.numeric(start) + as.numeric(end))/2) %>%
  mutate(chromo = chr) %>%
  merge(., chromo_cumpos, by = "chromo") %>%
  mutate(bpcum = MAPINFO + position_x) %>%
  as.data.frame()

mix_frail_dmr_stouffer$cate = factor(mix_frail_dmr_stouffer$cate, levels = c("frail", "dual", "age"))
ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = 0, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = mix_frail_dmr_stouffer,
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 size = log2(no.cpgs),
                 color = as.factor(cate),
                 shape = as.factor(cate))) +
  scale_size(range = c(1.5, 4)) +
  scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) + 
  scale_shape_manual(values = c(19, 17, 15)) +
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
  labs(x = "", y = "-log10(Stouffer)", color = "", shape = "", size = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust = 1, size = 12, face = "bold")) 
```

## 2.5. Examples for aiDMRs, dualDMRs and adDMRs
```{r}
methy_dat_dis_meta_excl = methy_dat_dis_meta %>% 
  as.data.frame() %>% 
  mutate(logFI = log2(FI.Score),
         mouse_id = as.factor(Mouse.ID)) %>%
  filter(Time.label != "T5")
methy_dat_dis_excl = methy_dat_dis %>% filter(rownames(.) %in% methy_dat_dis_meta_excl$id)
dmr1_visual = mix_loci_dmrs %>% filter(dmr == "chr12:49382107-49385959") %>% arrange(MAPINFO) 

dmr1_cpg_betas = methy_dat_dis_excl %>%
    dplyr::select(dmr1_visual$IlmnID) %>%
    mutate(id = rownames(.)) %>%
    merge(., methy_dat_dis_meta_excl %>% dplyr::select(id, logFI, mouse_id), by = "id") %>%
    as.data.frame() %>%
    melt(., id = c("id", "logFI", "mouse_id")) %>%
    as.data.frame() %>%
    merge(., dmr1_visual, by.x = "variable", by.y = "IlmnID") %>%
    as.data.frame()

dmr1_cpg_location_dat = data.frame(x1 = c(49382000, 49383078, 49385808),
                                   x2 = c(49383078, 49385808, 49386000),
                                   cate = c("shore", "island", "shore"))

dmr1_exons_location_dat = data.frame(x1 = c(49382000, 49382883),
                                     x2 = c(49382883, 49386000),
                                     cate = c("non", "exon"))

ggplot(data = dmr1_cpg_betas) +
  geom_rect(aes(xmin = 49382107, 
                xmax = 49385959, 
                ymin = 0, 
                ymax = Inf), 
            fill = "#f5f0f0",
            inherit.aes = FALSE) + 
  geom_rect(data = dmr1_cpg_location_dat,
            aes(xmin = x1,
                xmax = x2,
                ymin = -0.15,
                ymax = -0.1,
                fill = as.factor(cate)),
            inherit.aes = FALSE) +
  scale_fill_manual(values = c("#DDCC77", "#CC6677")) +
  new_scale_fill() +
  geom_rect(data = dmr1_exons_location_dat,
            aes(xmin = x1,
                xmax = x2,
                ymin = -0.2,
                ymax = -0.25,
                fill = as.factor(cate)),
            inherit.aes = FALSE) +
  scale_fill_manual(values = c("#774411", "white")) +
  geom_smooth(aes(x = MAPINFO,
                  y = value,
                  group = id,
                  color = logFI),
              method = "loess", span = 0.4, 
              se = FALSE, size = 0.4, alpha = 0.2) +
  geom_point(aes(x = MAPINFO,
                 y = value,
                 color = logFI),
             size = 0.75) +
  scale_color_bamako(midpoint = -2.5) +
  new_scale_color() +
  ylim(-0.28, 1) +
  labs(x = "Chromosome coordinate (Kbp)", y = "Beta values") +
  geom_hline(yintercept = 0) +
  scale_color_manual(values = c("#762A83")) +
  scale_x_continuous(limits = c(49382000, 49386000), expand = c(0.01, 0.01))
```

```{r}
dmr2_visual = mix_loci_dmrs %>% filter(dmr == "chr1:42693424-42699277") %>% arrange(MAPINFO) 
dmr2_cpg_betas = methy_dat_dis_excl %>%
    dplyr::select(dmr2_visual$IlmnID) %>%
    mutate(id = rownames(.)) %>%
    merge(., methy_dat_dis_meta_excl %>% dplyr::select(id, logFI, mouse_id), by = "id") %>%
    as.data.frame() %>%
    melt(., id = c("id", "logFI", "mouse_id")) %>%
    as.data.frame() %>%
    merge(., dmr2_visual, by.x = "variable", by.y = "IlmnID") %>%
    as.data.frame()

dmr2_cpg_location_dat = data.frame(x1 = c(42693400, 42694075, 42694484, 42694734, 42695261, 42696575, 42698657),
                                   x2 = c(42694075, 42694484, 42694734, 42695261, 42696575, 42698657, 42699300),
                                   cate = c("shore", "island", "shore", "island", "shore", "island", "shore"))

dmr2_exons_location_dat = data.frame(x1 = c(42693400, 42694747, 42694825, 42694908, 42695188, 42695768, 42699300),
                                     x2 = c(42694747, 42694825, 42694908, 42695188, 42695768, 42699300, 42699300),
                                     cate = c("non", "exon", "non", "exon", "non", "exon", "non"))

#pdf(file = "~/Desktop/DNA methylation/figures_0718/suppl_fig3f.pdf", width = 7, height = 4.2)
ggplot(data = dmr2_cpg_betas) +
  geom_rect(aes(xmin = 42693424, 
                xmax = 42699277, 
                ymin = 0, 
                ymax = Inf), 
            fill = "#f5f0f0",
            inherit.aes = FALSE) + 
  geom_rect(data = dmr2_cpg_location_dat,
            aes(xmin = x1,
                xmax = x2,
                ymin = -0.15,
                ymax = -0.1,
                fill = as.factor(cate)),
            inherit.aes = FALSE) +
  scale_fill_manual(values = c("#DDCC77", "#CC6677")) +
  new_scale_fill() +
  geom_rect(data = dmr2_exons_location_dat,
            aes(xmin = x1,
                xmax = x2,
                ymin = -0.2,
                ymax = -0.25,
                fill = as.factor(cate)),
            inherit.aes = FALSE) +
  scale_fill_manual(values = c("#774411", "white")) +
  geom_smooth(aes(x = MAPINFO,
                  y = value,
                  group = id,
                  color = logFI),
              method = "loess", span = 0.5, 
              se = FALSE, size = 0.4, alpha = 0.2) +
  geom_point(aes(x = MAPINFO,
                 y = value,
                 color = logFI),
             size = 0.75) +
  scale_color_bamako(midpoint = -2.5) +
  new_scale_color() +
  ylim(-0.3, 1) +
  labs(x = "Chromosome coordinate (Kbp)", y = "Beta values") +
  geom_hline(yintercept = 0) +
  scale_color_manual(values = c("#44AA99", "#DDAA33")) +
  scale_x_continuous(limits = c(42693400, 42699300), expand = c(0.01, 0.01))
```

## 2.6. enrichment analysis for CpG of AIFI and ADFI
```{r}
mix_aidualdmr = mix_dmr_assoc_res_adj %>% filter(cate %in% c("frail", "dual"))
mix_aidualdmr_cpgs = mix_loci_dmrs[mix_loci_dmrs$dmr %in% mix_aidualdmr$dmr, ]

mix_addualdmr = mix_dmr_assoc_res_adj %>% filter(cate %in% c("age", "dual"))
mix_addualdmr_cpgs = mix_loci_dmrs[mix_loci_dmrs$dmr %in% mix_addualdmr$dmr, ]

mix_fidmr = mix_dmr_assoc_res_adj %>% filter(cate != "other")
mix_fidmr_cpgs = mix_loci_dmrs[mix_loci_dmrs$dmr %in% mix_fidmr$dmr, ]

mix_aidualdmrcpg_genes = prob_enriched_genes(mix_aidualdmr_cpgs$IlmnID)
```

```{r}
suppl_tbl1 = merge(mix_fidmr_cpgs, mix_dmr_assoc_res_adj, by = "dmr") %>%
  as.data.frame() %>%
  dplyr::select(IlmnID, MAPINFO, dmr, cate) %>%
  arrange(cate)
write.csv(suppl_tbl1, file = "suppl_tbl1.csv")

suppl_tbl2 = mix_aidualdmrcpg_genes %>%
  dplyr::select(gene_name, dbname, estimate, FDR, nQ, nD, overlap)
write.csv(suppl_tbl2, file = "suppl_tbl2.csv")
```

```{r}
mix_aidualdmr_path = prob_enriched_path(mix_aidualdmr_cpgs$IlmnID)
mix_addualdmr_path = prob_enriched_path(mix_addualdmr_cpgs$IlmnID)
mix_fidmr_path = prob_enriched_path(mix_fidmr_cpgs$IlmnID)

mix_frailty_path = rbind(mix_aidualdmr_path %>% mutate(type = "AIFI"),
                           mix_addualdmr_path %>% mutate(type = "ADFI"),
                           mix_fidmr_path %>% mutate(type = "fi")) %>%
  as.data.frame()
mix_frailty_path = within(mix_frailty_path, 
                                 term_name <- ave(as.character(mix_frailty_path$term_name), 
                                                  FUN = make.unique))
mix_frailty_path$term_name = factor(mix_frailty_path$term_name, 
                                           levels = mix_frailty_path$term_name)

#pdf(file = "~/Desktop/DNA methylation/figures_0718/fig2c&d.pdf", width = 7, height = 5)
ggplot(data = mix_frailty_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point()+
    scale_size(range = c(1, 3)) +
    scale_color_bamako(reverse = TRUE) +
    facet_grid(rows = vars(type), 
             scales = "free_y", 
             space = "free") +
    theme(axis.text.y = element_text(size = 8, face = "bold")) +
    labs(y = "") 
```

## 2.7. addmr_cpgs_heatmap
```{r}
mix_age_plot_order = metadata_fm_excl %>%
                     arrange(., `Sex`,`Time.label`)
mix_comb_age_plot_dat = methy_dat_dis_M %>%
                     filter(rownames(.) %in% mix_age_plot_order$id) %>%
                     dplyr::select(mix_addualdmr_cpgs$IlmnID) %>%
                     dplyr::slice(order(match(rownames(.), mix_age_plot_order$id))) %>%
                     t() %>% as.matrix()

ComplexHeatmap::pheatmap(mix_comb_age_plot_dat,
         color = colorRampPalette(c("#DDCC77", "grey90", "#004488"))(20),
         scale = "row",
         cluster_cols = F,
         cluster_rows = T,
         annotation_col = data.frame(`Time_point` = mix_age_plot_order$Time.label,
                                     Sex = mix_age_plot_order$Sex),
         annotation_colors = list(`Time_point` = c(BL = "#332288", T2 = "#AAAA00", T3 = "#44AA99", 
                                                   T4 = "#882255"),
                                  Sex = c(female = "#AA4466", male = "#4477AA")),
         show_rownames = FALSE,
         use_raster = TRUE)
```

# 3. fi methylation markers in females and males

## 3.1. limma analysis for DMPs
```{r}
## sex-specific limma
metadata_male_only = methy_dat_dis_meta %>% filter(Sex == "male")
frail_male = log2(metadata_male_only$FI.Score)
design_male = model.matrix(~ frail_male)
colnames(design_male) = make.names(colnames(design_male))
male_dat_limma = methy_dat_dis_M %>% filter(rownames(.) %in% metadata_male_only$id)
male_dat_counts = male_dat_limma %>% t() %>% as.data.frame() 
male_residuals = residuals(maleonly_fit2_res, male_dat_counts)
maleonly_fit2_res = get_fit2_res(metadata_male_only, design_male, methy_dat_dis_M)
metadata_female_only = methy_dat_dis_meta %>% filter(Sex == "female")
frail_female = log2(metadata_female_only$FI.Score)
female_dat_limma = methy_dat_dis_M %>% filter(rownames(.) %in% metadata_female_only$id)
female_dat_counts = female_dat_limma %>% t() %>% as.data.frame() 
female_residuals = residuals(femaleonly_fit2_res, female_dat_counts)
design_female = model.matrix(~ frail_female)
colnames(design_female) = make.names(colnames(design_female))
femaleonly_fit2_res = get_fit2_res(metadata_female_only, design_female, methy_dat_dis_M)
```

```{r}
## female/male manhattan plot for DMPs
frail_male_tt = topTable(maleonly_fit2_res, coef = 2, n = Inf, adjust.method = "BH", sort.by = "none")
frail_male_dmp = frail_male_tt %>% filter(adj.P.Val < 0.05) # nrow 69091
frail_female_tt = topTable(femaleonly_fit2_res, coef = 2, n = Inf, adjust.method = "BH", sort.by = "none")
frail_female_dmp = frail_female_tt %>% filter(adj.P.Val < 0.05) # nrow 90401

female_male_tt = rbind(frail_male_tt %>% 
                         mutate(IlmnID = rownames(.), sex = "m"),
                       frail_female_tt %>% 
                         mutate(IlmnID = rownames(.), sex = "f")) %>%
  as.data.frame()
female_male_cpg_pos = merge(female_male_tt,
                  cpg_loci,
                  by = "IlmnID") %>% 
    as.data.frame() %>%
    dplyr::select(IlmnID, chromo, MAPINFO, P.Value, adj.P.Val, sex)
female_male_pos_cumpos = merge(female_male_cpg_pos, chromo_cumpos, by = "chromo") %>%
    mutate(bpcum = MAPINFO + position_x) %>%
    dplyr::select(chromo, IlmnID, P.Value, adj.P.Val, MAPINFO, position_x, bpcum, sex)
female_male_pos_cumpos$chromo = factor(female_male_pos_cumpos$chromo, levels = chrOrder_DNAm)
female_pos_cumpos_nonsig = female_male_pos_cumpos %>% filter(sex == "f") %>% 
  filter(chromo != "chrY") %>% filter(adj.P.Val >= 0.05)
female_pos_cumpos_sig = female_male_pos_cumpos %>% filter(sex == "f") %>% 
  filter(chromo != "chrY") %>% filter(adj.P.Val < 0.05)
male_pos_cumpos_nonsig = female_male_pos_cumpos %>% filter(sex == "m") %>% filter(adj.P.Val >= 0.05)
male_pos_cumpos_sig = female_male_pos_cumpos %>% filter(sex == "m") %>% filter(adj.P.Val < 0.05)
female_male_manhattan_plot = ggplot() +
    geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = -Inf, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
    geom_point(data = female_pos_cumpos_nonsig,
               aes(x = bpcum,
                   y = -log10(P.Value),
                   color = as.factor(chromo)), size = 0.2) + 
    scale_color_manual(values = rep(c("grey65", "grey80"), 11)) +
    guides(color = FALSE) + 
    new_scale_color() + 
    geom_point(data = female_pos_cumpos_sig,
               aes(x = bpcum,
                   y = -log10(P.Value), 
                   color = as.factor(chromo)), size = 0.4) + 
    scale_color_manual(values = rep(c("#AA4466", "#eba7b2"), 11)) +
    guides(color = FALSE) + 
    new_scale_color() +
    geom_point(data = male_pos_cumpos_nonsig,
               aes(x = bpcum,
                   y = log10(P.Value),
                   color = as.factor(chromo)), size = 0.2) + 
    scale_color_manual(values = rep(c("grey65", "grey80"), 11)) +
    guides(color = FALSE) + 
    new_scale_color() + 
    geom_point(data = male_pos_cumpos_sig,
               aes(x = bpcum,
                   y = log10(P.Value), 
                   color = as.factor(chromo)), size = 0.4) + 
    scale_color_manual(values = rep(c("#4477AA", "#90b1d1"), 11)) +
    guides(color = FALSE) + 
    scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
    labs(x = "", y = "-log10(p-value)", color = "") +
    geom_hline (yintercept = 0, lty = 3) + 
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 12, face = "bold"))
rasterize(female_male_manhattan_plot, dpi = 300)
dev.off()
```

```{r}
#mix_all_anno = cpg_dat_context_anno %>%
 # filter(IlmnID %in% mix_tt_cpg_loci$IlmnID)
female_tt_cpg_loci = merge(frail_female_tt %>% mutate(IlmnID = rownames(.)),
                        cpg_loci,
                        by = "IlmnID") %>%
  as.data.frame()
male_tt_cpg_loci = merge(frail_male_tt %>% mutate(IlmnID = rownames(.)),
                        cpg_loci,
                        by = "IlmnID") %>%
  as.data.frame()
female_tt_cpg_loci_dmp = merge(frail_female_tt %>% mutate(IlmnID = rownames(.)),
                        cpg_loci,
                        by = "IlmnID") %>%
  filter(adj.P.Val < 0.05)
male_tt_cpg_loci_dmp = merge(frail_male_tt %>% mutate(IlmnID = rownames(.)),
                        cpg_loci,
                        by = "IlmnID") %>%
  filter(adj.P.Val < 0.05)
dmps_anno_cpg_lst = list(
  female_tt_cpg_loci_dmp$IlmnID,
  female_tt_cpg_loci_dmp[!female_tt_cpg_loci_dmp$chromo %in% c("chrX", "chrY", "chrMT"), ]$IlmnID,
  female_tt_cpg_loci_dmp[female_tt_cpg_loci_dmp$chromo == "chrX", ]$IlmnID,
  male_tt_cpg_loci_dmp$IlmnID,
  male_tt_cpg_loci_dmp[!male_tt_cpg_loci_dmp$chromo %in% c("chrX", "chrY", "chrMT"), ]$IlmnID,
  male_tt_cpg_loci_dmp[male_tt_cpg_loci_dmp$chromo == "chrX", ]$IlmnID)
fm_dmp_anno = lapply(1:6, function(x) {
  dmp_anno = cpg_dat_context_anno %>%
    filter(IlmnID %in% dmps_anno_cpg_lst[[x]]) %>%
    mutate(type = x)
}) %>%
  Reduce(rbind, .) %>%
  rbind(., mix_all_anno %>% mutate(type = "all")) %>%
  as.data.frame()

fm_dmp_anno_cpg_dmp = fm_dmp_anno %>%
  dplyr::group_by(type, CpG_context) %>%
  dplyr::tally() %>%
  dplyr::mutate(prop_cpg = n/sum(n))

fm_dmp_anno_gene_dmp = fm_dmp_anno %>%
  dplyr::group_by(type, Gene_context) %>%
  dplyr::tally() %>%
  dplyr::mutate(prop_gene = n/sum(n))

fm_dmp_anno_cpg_dmp$CpG_context = factor(fm_dmp_anno_cpg_dmp$CpG_context, 
                                          levels = c("Island", "Shore", "Shelf", "Open Sea"))
fm_dmp_anno_cpg_dmp$type = factor(fm_dmp_anno_cpg_dmp$type,
                                   levels = c("all", "1", "2", "3", "4", "5", "6"))
fm_dmp_anno_cpg_context_plot = ggplot(data = fm_dmp_anno_cpg_dmp,
       aes(x = type,
           y = prop_cpg,
           fill = CpG_context)) +
  geom_bar(stat = "identity", position = "fill") +
   labs(x = "") +
  scale_fill_manual(values = c("#DDCC77", "#CC6677", "#AA4499", "#882255")) +
  scale_x_discrete(label = c("All CpGs", "F. DMPs", "F. DMPs (at.)", "F. DMPs (chrX)", 
                             "M. DMPs", "M. DMPs (at.)", "M. DMPs (chrX)")) +
  theme(axis.text.x = element_text(angle = 60, vjust = 0.5, hjust=1, size = 4, face = "bold"))

fm_dmp_anno_gene_dmp$Gene_context = factor(fm_dmp_anno_gene_dmp$Gene_context, 
                                              levels = c("tss_200", "tss_1500", "tss_body", "Intergenic"))
fm_dmp_anno_gene_dmp$type = factor(fm_dmp_anno_gene_dmp$type,
                                   levels = c("all", "1", "2", "3", "4", "5", "6"))
fm_dmp_anno_gene_context_plot = ggplot(data = fm_dmp_anno_gene_dmp,
       aes(x = type,
           y = prop_gene,
           fill = Gene_context)) +
  geom_bar(stat = "identity", position = "fill") +
  labs(x = "") +
  scale_fill_manual(values = c("#DDCC77", "#CC6677", "#AA4499", "#882255" )) +
  scale_x_discrete(label = c("All CpGs", "F. DMPs", "F. DMPs (at.)", "F. DMPs (chrX)", 
                             "M. DMPs", "M. DMPs (at.)", "M. DMPs (chrX)")) +
  theme(axis.text.x = element_text(angle = 60, vjust = 0.5, hjust=1, size = 4, face = "bold"))

ggarrange(fm_dmp_anno_cpg_context_plot, fm_dmp_anno_gene_context_plot,
          nrow = 1, ncol = 2) 
```

## 3.2 XCI analysis 
```{r}
cpg_chrx = cpg_pos_cumpos %>% filter(chromo == "chrX") %>%
  arrange(MAPINFO)
methy_dat_cpg_chrx_melt = methy_dat_dis %>% 
                    dplyr::select(all_of(cpg_chrx$IlmnID)) %>%
                    mutate(id = rownames(.),
                           sex = methy_dat_dis_meta$Sex) %>% 
                    melt(., id = c("id", "sex")) %>% as.data.frame() 

ggplot(data = methy_dat_cpg_chrx_melt,
       aes(x = value,
           color = sex)) +
    geom_density(linewidth = 1.25) +
    scale_color_manual(values = c("#AA4466", "#4477AA"))
```

```{r}
female_dat = methy_dat_dis_meta %>% filter(Sex == "female")
male_dat = methy_dat_dis_meta %>% filter(Sex == "male")
methy_dat_cpg_chrx = methy_dat_dis %>% 
                    dplyr::select(all_of(cpg_chrx$IlmnID)) 
methy_dat_cpg_chrx_female = methy_dat_dis %>% 
                    dplyr::select(all_of(cpg_chrx$IlmnID)) %>%
                    filter(rownames(.) %in% female_dat$id)
methy_dat_cpg_chrx_male = methy_dat_dis %>% 
                    dplyr::select(all_of(cpg_chrx$IlmnID)) %>%
                    filter(rownames(.) %in% male_dat$id)
female_ave = colSums(methy_dat_cpg_chrx_female)/nrow(methy_dat_cpg_chrx_female)
male_ave = colSums(methy_dat_cpg_chrx_male)/nrow(methy_dat_cpg_chrx_male)

check_range_over = function(min_f, max_f, min_m, max_m) {
  if (min_m <= max_f & min_m >= min_f) {
    over = "yes"
  } else if (max_m <= max_f & max_m >= min_f) {
    over = "yes"
  } else {over = "no"}
  return(over)
}

methy_dat_cpg_chrx_female_range = lapply(cpg_chrx$IlmnID, function(x) {
  cpg_df = methy_dat_cpg_chrx_female[, x]
  return(c(min(cpg_df), max(cpg_df)))
}) %>%
  Reduce(rbind, .) %>%
  as.data.frame() %>%
  `colnames<-`(c("min_f", "max_f")) %>%
  mutate(IlmnID = cpg_chrx$IlmnID)
  
methy_dat_cpg_chrx_male_range = lapply(cpg_chrx$IlmnID, function(x) {
  cpg_df = methy_dat_cpg_chrx_male[, x]
  return(c(min(cpg_df), max(cpg_df)))
}) %>%
  Reduce(rbind, .) %>%
  as.data.frame() %>%
  `colnames<-`(c("min_m", "max_m")) %>%
  mutate(IlmnID = cpg_chrx$IlmnID)

range_over = sapply(1:length(cpg_chrx$IlmnID), function(x) {
  min_f = methy_dat_cpg_chrx_female_range[x, ]$min_f
  max_f = methy_dat_cpg_chrx_female_range[x, ]$max_f
  min_m = methy_dat_cpg_chrx_male_range[x, ]$min_m
  max_m = methy_dat_cpg_chrx_male_range[x, ]$max_m
  over = check_range_over(min_f, max_f, min_m, max_m)
  return(over)
})

fvsm_chrx_beta = data.frame(IlmnID = cpg_chrx$IlmnID,
                            female = female_ave,
                            male = male_ave,
                            diff = female_ave - male_ave) %>%
  merge(., cpg_chrx, by = "IlmnID") %>%
  merge(., methy_dat_cpg_chrx_female_range, by = "IlmnID") %>%
  merge(., methy_dat_cpg_chrx_male_range, by = "IlmnID") %>%
  as.data.frame() %>%
  mutate(over = range_over)

assign_Xci = function(female, male, over, diff) {
  if (female < 0.15 & male < 0.15) {
    if (over == "yes") {
      if (diff < 0.1) {
        xci = "escape"
      } else {xci = "variable escape"}
    } else {
      if (diff < 0.1) {
        xci = "escape"
      } else {xci = "XCI"}
    }
  } else {
    if (over == "yes") {
      if (diff < 0.1) {
        xci = "unclassifiable"
      } else {xci = "variable escape"}
    } else {
      if (diff < 0.1) {
        xci = "unclassifiable"
      } else {xci = "XCI"}
    }
  }
  return(xci)  
}

chrx_cpg_xci = sapply(1:length(cpg_chrx$IlmnID), function(x) {
  female = fvsm_chrx_beta[x, ]$female
  male = fvsm_chrx_beta[x, ]$male
  over = fvsm_chrx_beta[x, ]$over
  diff = fvsm_chrx_beta[x, ]$diff
  xci_status = assign_Xci(female, male, over, diff)
  return(xci_status)
})

fvsm_chrx_beta_update = fvsm_chrx_beta %>% 
  mutate(xci_status = chrx_cpg_xci)
fvsm_chrx_beta_update$xci_status = factor(fvsm_chrx_beta_update$xci_status, 
                                          levels = c("variable escape", "unclassifiable", "XCI", "escape"))

ggplot() +
  geom_hline(yintercept = 0, lty = 3, color = "grey40", linewidth = 1.25) +
  geom_point(data = fvsm_chrx_beta_update %>% 
               filter(xci_status %in% c("variable escape", "unclassifiable")),
             aes(x = MAPINFO,
                 y = diff,
                 color = xci_status),
             size = 0.3) +
  scale_color_manual(values = c("#999933", "#DDAA33")) +  
  new_scale_color() +
  geom_point(data = fvsm_chrx_beta_update %>% 
               filter(xci_status %in% c("XCI", "escape")),
             aes(x = MAPINFO,
                 y = diff,
                 color = xci_status,
                 shape = xci_status,
                 size = xci_status),
             size = 1.8) +
   xlim(5468000, 170677400) +
   scale_color_manual(values = c("#762A83", "#44AA99")) +
   scale_shape_manual(values = c(18, 15)) +
   scale_size_manual(values = c(0.8, 2.5)) +
   scale_x_continuous(breaks = c(54000000, seq(5400000, 171000000, 20700000)),
        expand = c(0.01, 0.01))
```

```{r}
fm_xci_cpg = rbind(female_tt_cpg_loci %>% filter(chromo == "chrX") %>% filter(adj.P.Val < 0.05) %>% mutate(sex = "f"),
                   male_tt_cpg_loci %>% filter(chromo == "chrX") %>% filter(adj.P.Val < 0.05) %>% mutate(sex = "m")) %>%
  as.data.frame() %>%
  dplyr::select(IlmnID, sex) %>%
  merge(., fvsm_chrx_beta_update, by = "IlmnID") %>%
  as.data.frame() 

female_XCA_genes = fm_xci_cpg %>% filter(sex == "f" & xci_status == "escape")
male_XCA_genes = fm_xci_cpg %>% filter(sex == "m" & xci_status == "escape")
female_specific_XCA = setdiff(female_XCA_genes$IlmnID, intersect(female_XCA_genes$IlmnID, male_XCA_genes$IlmnID))

#prob_enriched_genes(female_specific_XCA)
```

## 3.3. Female specific analysis for fi methylation makers
```{r}
female_cpg_fc = merge(frail_female_tt %>% mutate(IlmnID = rownames(.)), 
                      mix_cpg_loci, by = "IlmnID")
female_dmr_all = get_dmr_modify_DMRcate(female_cpg_fc) 
female_dmr_greater5 = female_dmr_all %>% filter(no.cpgs >= 5) %>%
  filter(chr != "chrY")

frail_female_dmr = female_dmr_greater5 %>% filter(Stouffer < 0.05)
frail_female_loci_dmrs = select_dmp_for_dmr(frail_female_dmr)

female_dmr_stouffer = female_dmr_greater5 %>%
  dplyr::select(coord, no.cpgs, chr, start, end, Stouffer) %>%
  as.data.frame() %>%
  mutate(MAPINFO = (as.numeric(start) + as.numeric(end))/2) %>%
  mutate(chromo = chr) %>%
  merge(., chromo_cumpos, by = "chromo") %>%
  mutate(bpcum = MAPINFO + position_x) 

ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = 0, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = female_dmr_stouffer %>% filter(Stouffer < 0.05),
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 size = log2(no.cpgs),
                 color = as.factor(chromo))) +
  scale_size(range = c(1, 3)) +
  scale_color_manual(values = c(rep(c("#114477", "#44AA99"), 10), "#44AA99")) +
  guides(color = "none") +
  new_scale_color() +
  geom_point(data = female_dmr_stouffer %>% filter(Stouffer >= 0.05),
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 color = as.factor(chromo)),
             size = 0.2) +
  scale_color_manual(values = c(rep(c("grey65", "grey80"), 10), "grey80")) +
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
  labs(x = "", y = "-log10(Stouffer)", color = "", ) + guides(color = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 9, face = "bold")) 
```

## 3.3. coMethyR analysis in females
```{r}
female_dmps_expr = methy_dat_dis_M %>% filter(rownames(.) %in% metadata_female_only$id) %>%
                      dplyr::select(unlist(frail_female_loci_dmrs$IlmnID))
female_dmps_expr_merged = merge(female_dmps_expr %>% mutate(id = rownames(.)), 
                metadata_female_only, by = "id") %>%
                  as.data.frame()
female_test_rdrop = test_rdrop(frail_female_loci_dmrs, female_dmps_expr)
female_rdrop_loci_remain_v1 = female_test_rdrop %>% filter(rdrop > 0.4)
female_rdrop_dmr_above5 = table(female_rdrop_loci_remain_v1$dmr) %>% as.data.frame() %>% filter(Freq >= 5)
female_rdrop_loci_remain_v2 = female_rdrop_loci_remain_v1 %>% filter(dmr %in% female_rdrop_dmr_above5$Var1)
```

```{r, warning = FALSE, message = FALSE}
female_loci_expr_assoc = female_dmps_expr_merged %>% as.data.frame() %>%
  mutate(age = `Age.at.assesment..days.`,
         log2FI = log2(FI.Score))

female_dmr_assoc_res = lapply(unique(female_rdrop_loci_remain_v2$dmr), function(x) {
  loci_lmer = female_rdrop_loci_remain_v2 %>% filter(dmr == x)
  expr_merged = female_loci_expr_assoc %>%
    dplyr::select(loci_lmer$IlmnID, id, Mouse.ID, age, log2FI) %>%
    melt(., id = c("id", "Mouse.ID", "age", "log2FI")) %>%
    as.data.frame()
  res1 = lmer(value ~ log2FI + age + (1 | Mouse.ID) + (log2FI | variable), data = expr_merged)
  res2 = lmer(value ~ age + (1 | Mouse.ID) + (age | variable), data = expr_merged)
  return(c(summary(res1)$coefficients[2, 5], summary(res2)$coefficients[2, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame()
rownames(female_dmr_assoc_res) = unique(female_rdrop_loci_remain_v2$dmr)
colnames(female_dmr_assoc_res) = c("p_frail", "p_age")

female_dmr_assoc_res_adj = female_dmr_assoc_res %>%
  mutate(adjp_frail = p.adjust(p_frail, method = "fdr"),
         adjp_age = p.adjust(p_age, method = "fdr")) %>%
  mutate(dmr_cate = case_when(
    adjp_frail < 0.05 & adjp_age < 0.05 ~ "dual",
    adjp_age < 0.05 ~ "age",
    adjp_frail < 0.05 ~ "frail",
    TRUE ~ "other"
  )) %>% 
  mutate(dmr = rownames(.))

female_dmr_assoc_res_adj$dmr_cate = factor(female_dmr_assoc_res_adj$dmr_cate, 
                                           levels = c("dual", "age", "other"))

ggplot(data = female_dmr_assoc_res_adj,
       aes(x = -log10(adjp_frail),
           y = -log10(adjp_age),
           color = as.factor(dmr_cate),
           shape = as.factor(dmr_cate))) + 
  geom_point(size = 1.5) +
  geom_hline(yintercept = -log10(0.05), lty = 3, color = "grey50") +
  geom_vline(xintercept = -log10(0.05), lty = 3, color = "grey50") +
  scale_color_manual(values = c("#DDAA33", "#44AA99", "#999933")) +
  scale_shape_manual(values = c(17, 15, 18)) + 
  expand_limits(x = 0, y = 0) +
  labs(x = "-log10(p)_frail", 
       y = "-log10(p)_age",
       color = "")
```

```{r}
## examples of DMR
methy_dat_dis_meta_f = methy_dat_dis_meta %>% 
  as.data.frame() %>% 
  mutate(logFI = log2(FI.Score),
         mouse_id = as.factor(Mouse.ID)) %>%
  filter(Sex == "female")
methy_dat_dis_f = methy_dat_dis %>% filter(rownames(.) %in% methy_dat_dis_meta_f$id)
f_dmr_visual = frail_female_loci_dmrs %>% 
  filter(dmr == "chr2:174325146-174329029") %>%
  arrange(MAPINFO)

f_dmr_cpg_betas = methy_dat_dis_f %>%
    dplyr::select(f_dmr_visual$IlmnID) %>%
    mutate(id = rownames(.)) %>%
    merge(., methy_dat_dis_meta_f %>% dplyr::select(id, logFI, mouse_id), by = "id") %>%
    as.data.frame() %>%
    melt(., id = c("id", "logFI", "mouse_id")) %>%
    as.data.frame() %>%
    merge(., f_dmr_visual, by.x = "variable", by.y = "IlmnID") %>%
    as.data.frame()

f_dmr_cpg_location_dat = data.frame(x1 = c(174325000, 174325294, 174327294),
                                   x2 = c(174325294, 174327294, 174329200),
                                   cate = c("shelf", "shore", "island"))

f_dmr_exons_location_dat = data.frame(x1 = c(174325000, 174325894, 174326150, 174327146, 174327387, 174327749, 174327941, 174328108),
                                     x2 = c(174325894, 174326150, 174327146, 174327387, 174327749, 174327941, 174328108, 174329200),
                                     cate = c("non", "exon", "non", "exon", "non", "exon", "non", "exon"))

ggplot(data = f_dmr_cpg_betas) +
  geom_rect(aes(xmin = 174325146, 
                xmax = 174329029, 
                ymin = 0, 
                ymax = Inf), 
            fill = "#f5f0f0",
            inherit.aes = FALSE) +
  geom_rect(data = f_dmr_cpg_location_dat,
            aes(xmin = x1,
                xmax = x2,
                ymin = -0.15,
                ymax = -0.1,
                fill = as.factor(cate)),
            inherit.aes = FALSE) +
  scale_fill_manual(values = c("#DDCC77", "#CC6677", "#AA4499")) +
  new_scale_fill() +
  geom_rect(data = f_dmr_exons_location_dat,
            aes(xmin = x1,
                xmax = x2,
                ymin = -0.2,
                ymax = -0.25,
                fill = as.factor(cate)),
            inherit.aes = FALSE) +
  scale_fill_manual(values = c("#774411", "white")) +
  geom_smooth(aes(x = MAPINFO,
                  y = value,
                  group = id,
                  color = logFI),
              method = "loess", span = 0.7, 
              se = FALSE, size = 0.4, alpha = 0.2) +
  geom_point(aes(x = MAPINFO,
                 y = value,
                 color = logFI),
             size = 0.75) +
  scale_color_bamako(midpoint = -2.5) +
  new_scale_color() +
  ylim(-0.3, 1) +
  labs(x = "", y = "") +
  labs(x = "Chromosome coordinate (Kbp)", y = "Beta values") +
  geom_hline(yintercept = 0) +
  scale_x_continuous(expand = c(0.01, 0.01))
```

```{r}
female_frail_dmr_stouffer = female_dmr_greater5 %>%
  dplyr::select(coord, no.cpgs, chr, start, end, Stouffer) %>%
  mutate(MAPINFO = (as.numeric(start) + as.numeric(end))/2) %>%
  mutate(chromo = chr) %>%
  merge(., chromo_cumpos, by = "chromo") %>%
  mutate(bpcum = MAPINFO + position_x) %>%
  merge(., female_dmr_assoc_res_adj, by.x = "coord", by.y = "dmr") %>%
  as.data.frame() %>%
  filter(dmr_cate != "other")

female_frail_dmr_stouffer$dmr_cate = factor(female_frail_dmr_stouffer$dmr_cate, levels = c("dual", "age"))
ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = 0, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = female_frail_dmr_stouffer,
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 size = log2(no.cpgs),
                 color = as.factor(dmr_cate),
                 shape = as.factor(dmr_cate))) +
  scale_size(range = c(1, 3)) +
  scale_color_manual(values = c("#DDAA33", "#44AA99")) + 
  scale_shape_manual(values = c(17, 15)) +
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
  labs(x = "", y = "-log10(Stouffer)", color = "", shape = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 12, face = "bold")) 
```

## 3.4 female fidmr enrichment analysis
```{r}
female_dualdmrs = female_dmr_assoc_res_adj[female_dmr_assoc_res_adj$dmr_cate == "dual", ]
female_dualdmrs_cpgs = frail_female_loci_dmrs[frail_female_loci_dmrs$dmr %in% rownames(female_dualdmrs), ]
mix_aidualdmrcpg_genes = prob_enriched_genes(female_dualdmrs_cpgs$IlmnID)
female_addmrs = female_dmr_assoc_res_adj[female_dmr_assoc_res_adj$dmr_cate == "age", ]
female_addmrs_cpgs = frail_female_loci_dmrs[frail_female_loci_dmrs$dmr %in% rownames(female_addmrs), ]
female_fidmrs = female_dmr_assoc_res_adj %>% filter(dmr_cate != "other")
female_fidmrs_cpgs = frail_female_loci_dmrs[frail_female_loci_dmrs$dmr %in% female_fidmrs$dmr, ]
```

```{r}
suppl_tbl3 = merge(female_fidmrs_cpgs, female_dmr_assoc_res_adj, by = "dmr") %>%
  as.data.frame() %>%
  dplyr::select(IlmnID, MAPINFO, dmr, dmr_cate) %>%
  arrange(dmr_cate)
write.csv(suppl_tbl3, file = "suppl_tbl3.csv")

suppl_tbl4 = mix_aidualdmrcpg_genes %>%
  dplyr::select(gene_name, dbname, estimate, FDR, nQ, nD, overlap)
write.csv(suppl_tbl4, file = "suppl_tbl4.csv")
```

```{r}
female_fidmrs_paths = prob_enriched_path(female_fidmrs_cpgs$IlmnID)
female_fidmrs_paths$term_name = factor(female_fidmrs_paths$term_name, levels = female_fidmrs_paths$term_name)
ggplot(data = female_fidmrs_paths,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_size(range = c(3, 5)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 8, face = "bold")) +
    labs(y = "")
```

## 3.6. Male specific analysis for fi methylation makers  
```{r}
male_cpg_fc = merge(frail_male_tt %>% mutate(IlmnID = rownames(.)), 
                      mix_cpg_loci, by = "IlmnID")
male_dmr_all = get_dmr_modify_DMRcate(male_cpg_fc) 
male_dmr_greater5 = male_dmr_all %>% filter(no.cpgs >= 5)

frail_male_dmr = male_dmr_greater5 %>% filter(Stouffer < 0.05)
frail_male_loci_dmrs = select_dmp_for_dmr(frail_male_dmr)

male_dmr_stouffer = male_dmr_greater5 %>%
  dplyr::select(coord, no.cpgs, chr, start, end, Stouffer) %>%
  as.data.frame() %>%
  mutate(MAPINFO = (as.numeric(start) + as.numeric(end))/2) %>%
  mutate(chromo = chr) %>%
  merge(., chromo_cumpos, by = "chromo") %>%
  mutate(bpcum = MAPINFO + position_x) 
ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = 0, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = male_dmr_stouffer %>% filter(Stouffer < 0.05),
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 size = log2(no.cpgs),
                 color = as.factor(chromo))) +
  scale_size(range = c(1, 3)) +
  scale_color_manual(values = rep(c("#114477", "#44AA99"), 11)) +
  guides(color = "none") +
  new_scale_color() +
  geom_point(data = male_dmr_stouffer %>% filter(Stouffer >= 0.05),
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 color = as.factor(chromo)),
             size = 0.2) +
  scale_color_manual(values = rep(c("grey65", "grey80"), 11)) +
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
  labs(x = "", y = "-log10(Stouffer)", color = "", ) + guides(color = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 12, face = "bold")) 
```

## 3.6. coMethyR analysis in males
```{r}
male_loci_dmr_expr = methy_dat_dis_M %>% filter(rownames(.) %in% metadata_male_only$id) %>%
                      dplyr::select(unlist(frail_male_loci_dmrs$IlmnID))
male_loci_dmr_expr_merged = merge(male_loci_dmr_expr %>% mutate(id = rownames(.)), 
                metadata_male_only, by = "id") %>%
                  as.data.frame()
male_test_rdrop = test_rdrop(frail_male_loci_dmrs, male_loci_dmr_expr)
male_rdrop_loci_remain_v1 = male_test_rdrop %>% filter(rdrop > 0.4)
male_rdrop_dmr_above5 = table(male_rdrop_loci_remain_v1$dmr) %>% as.data.frame() %>% filter(Freq >=5)
male_rdrop_loci_remain_v2 = male_rdrop_loci_remain_v1 %>% filter(dmr %in% male_rdrop_dmr_above5$Var1)
```

```{r, warning = FALSE, message = FALSE}
male_loci_dmr_expr_assoc = male_loci_dmr_expr_merged %>%
  mutate(age = `Age.at.assesment..days.`,
         log2FI = log2(FI.Score))
male_dmr_assoc_res = lapply(unique(male_rdrop_loci_remain_v2$dmr), function(x) {
  loci_lmer = male_rdrop_loci_remain_v2 %>% filter(dmr == x)
  expr_merged = male_loci_dmr_expr_assoc %>%
    dplyr::select(loci_lmer$IlmnID, id, Mouse.ID, age, log2FI) %>%
    melt(., id = c("id", "Mouse.ID", "age", "log2FI")) %>%
    as.data.frame()
  res1 = lmer(value ~ log2FI + age + (1 | Mouse.ID) + (log2FI | variable), data = expr_merged)
  res2 = lmer(value ~ age + (1 | Mouse.ID) + (age | variable), data = expr_merged)
  return(c(summary(res1)$coefficients[2, 5], summary(res2)$coefficients[2, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame()
rownames(male_dmr_assoc_res) = unique(male_rdrop_loci_remain_v2$dmr)
colnames(male_dmr_assoc_res) = c("p_frail", "p_age")
male_dmr_assoc_res_adj = male_dmr_assoc_res %>%
  mutate(adjp_frail = p.adjust(p_frail, method = "fdr"),
         adjp_age = p.adjust(p_age, method = "fdr")) %>%
  mutate(cate = case_when(
    adjp_frail < 0.05 & adjp_age < 0.05 ~ "dual",
    adjp_frail < 0.05 ~ "frail",
    adjp_age < 0.05 ~ "age",
    TRUE ~ "other")) %>%
  mutate(dmr = rownames(.))
```

```{r}
male_dmr_assoc_res_adj$cate = factor(male_dmr_assoc_res_adj$cate, levels = c("frail", "dual", "age", "other"))
#pdf(file = "~/Desktop/DNA methylation/figures_0718/suppl_fig5e.pdf", height = 2.1, width = 3)
ggplot(data = male_dmr_assoc_res_adj,
       aes(x = -log10(adjp_frail),
           y = -log10(adjp_age),
           color = as.factor(cate),
           shape = as.factor(cate))) + 
  geom_point(size = 1.5) +
  geom_hline(yintercept = -log10(0.05), lty = 3, color = "grey50") +
  geom_vline(xintercept = -log10(0.05), lty = 3, color = "grey50") +
  scale_color_manual(values = c("#762A83", "#44AA99", "#DDAA33", "#999933")) + 
  scale_shape_manual(values = c(19, 17, 15, 18)) +
  expand_limits(x = 0, y = 0) +
  labs(x = "-log10(p)_frail", 
       y = "-log10(p)_age",
       color = "")
```

```{r}
male_frail_dmr_stouffer = male_dmr_greater5 %>%
  dplyr::select(coord, no.cpgs, chr, start, end, Stouffer) %>%
  mutate(MAPINFO = (as.numeric(start) + as.numeric(end))/2) %>%
  mutate(chromo = chr) %>%
  merge(., chromo_cumpos, by = "chromo") %>%
  mutate(bpcum = MAPINFO + position_x) %>%
  merge(., male_dmr_assoc_res_adj, by.x = "coord", by.y = "dmr") %>%
  as.data.frame() %>%
  filter(cate != "other")

male_frail_dmr_stouffer$cate = factor(male_frail_dmr_stouffer$cate, levels = c("frail", "dual", "age"))
ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = -Inf, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = male_frail_dmr_stouffer,
             aes(x = bpcum,
                 y = -log10(Stouffer),
                 size = log2(no.cpgs),
                 color = as.factor(cate),
                 shape = as.factor(cate))) +
  scale_size(range = c(1, 3)) +
  scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) + 
  scale_shape_manual(values = c(19, 17, 15)) +
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, expand = c(0.01, 0.01)) +
  labs(x = "", y = "-log10(Stouffer)", color = "") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 12, face = "bold")) 
```

```{r}
male_aidualdmrs = male_dmr_assoc_res_adj %>% filter(cate %in% c("frail", "dual"))
male_aidualdmrs_cpgs = frail_male_loci_dmrs[frail_male_loci_dmrs$dmr %in% male_aidualdmrs$dmr, ]
male_addualdmrs = male_dmr_assoc_res_adj %>% filter(cate %in% c("age", "dual"))
male_addualdmrs_cpgs = frail_male_loci_dmrs[frail_male_loci_dmrs$dmr %in% male_addualdmrs$dmr, ]
male_fidmrs = male_dmr_assoc_res_adj %>% filter(cate != "other")
male_fidmrs_cpgs = frail_male_loci_dmrs[frail_male_loci_dmrs$dmr %in% male_fidmrs$dmr, ]  

male_aidualdmrs_genes = prob_enriched_genes(male_aidualdmrs_cpgs$IlmnID)
```

```{r}
male_aidualdmrs_path = prob_enriched_path(male_aidualdmrs_cpgs$IlmnID)
male_addualdmrs_path = prob_enriched_path(male_addualdmrs_cpgs$IlmnID)
male_fidmrs_path = prob_enriched_path(male_fidmrs_cpgs$IlmnID)

male_frailty_path = rbind(male_aidualdmrs_path %>% mutate(type = "AIFI"),
                            male_addualdmrs_path %>% mutate(type = "ADFI"),
                            male_fidmrs_path %>% mutate(type = "fi")) %>%
  as.data.frame()
male_frailty_path = within(male_frailty_path, 
                                 term_name <- ave(as.character(male_frailty_path$term_name), 
                                                  FUN = make.unique))
male_frailty_path$term_name = factor(male_frailty_path$term_name, 
                                           levels = male_frailty_path$term_name)

ggplot(data = male_frailty_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point()+
    scale_size(range = c(1, 3)) +
    scale_color_bamako(reverse = TRUE) +
    facet_grid(rows = vars(type), 
             scales = "free_y", 
             space = "free") +
    theme(axis.text.y = element_text(size = 8, face = "bold")) +
    labs(y = "") 
```

## 3.8. comparisons of fiCpGs between females and males
```{r}
male_aidmrs = male_dmr_assoc_res_adj %>% filter(cate == "frail")
male_dualdmrs = male_dmr_assoc_res_adj %>% filter(cate == "dual")
male_addmrs = male_dmr_assoc_res_adj %>% filter(cate == "age")
fm_loci_comp_list = list(`female_dualdmr` = female_dualdmrs$dmr,
                         `female_addmr` = female_addmrs$dmr,
                         `male_aidmr` = male_aidmrs$dmr,
                         `male_dualdmr` = male_dualdmrs$dmr,
                         `male_addmr` = male_addmrs$dmr)
m = make_comb_mat(fm_loci_comp_list)

ComplexHeatmap::UpSet(m,
      comb_col = c("#DDAA33", "#DDAA33", "#DDAA33", "#AA4466", "#AA4466", "#4477AA", "#4477AA", "#4477AA"), 
      set_order = c("female_dualdmr", "female_addmr", "male_aidmr", "male_dualdmr", "male_addmr"),
      pt_size = unit(5, "mm"),
      lwd = 3,
      top_annotation = upset_top_annotation(m, add_numbers = TRUE,
                                            numbers_gp = gpar(fontsize = 10, fontface = "bold"),
                                            numbers_rot = 0))
```

```{r}
fm_loci_comp_list = list(`female_addmr` = female_addmrs$dmr,
                         `male_aidmr` = male_aidmrs$dmr,
                         `male_dualdmr` = male_dualdmrs$dmr,
                         `male_addmr` = male_addmrs$dmr)
m = make_comb_mat(fm_loci_comp_list)

ComplexHeatmap::UpSet(m,
      comb_col = c("#DDAA33", "#DDAA33", "#DDAA33", "#AA4466", "#4477AA", "#4477AA", "#4477AA"), 
      set_order = c("female_addmr", "male_aidmr", "male_dualdmr", "male_addmr"),
      pt_size = unit(5, "mm"),
      lwd = 3,
      top_annotation = upset_top_annotation(m, add_numbers = TRUE,
                                            numbers_gp = gpar(fontsize = 10, fontface = "bold"),
                                            numbers_rot = 0))
```

```{r}
methy_dat_dis_meta_f = methy_dat_dis_meta %>% filter(Sex == "female")
methy_dat_dis_meta_m = methy_dat_dis_meta %>% filter(Sex == "male")
methy_dat_dis_M_f = methy_dat_dis_M %>% filter(rownames(.) %in% methy_dat_dis_meta_f$id) %>% as.data.frame()
methy_dat_dis_M_m = methy_dat_dis_M %>% filter(rownames(.) %in% methy_dat_dis_meta_m$id) %>% as.data.frame()
fm_all_cpgs = union(female_fidmrs_cpgs$IlmnID, male_fidmrs_cpgs$IlmnID)
fm_all_cpgs_df = rbind(female_fidmrs_cpgs, male_fidmrs_cpgs) %>%
  as.data.frame() %>%
  distinct()
fm_all_dmrs = union(unique(female_fidmrs_cpgs$dmr), unique(male_fidmrs_cpgs$dmr))
fm_inter_dmr = intersect(unique(female_fidmrs_cpgs$dmr), unique(male_fidmrs_cpgs$dmr))
```

```{r}
female_common_dmrs = female_dmr_assoc_res_adj %>% filter(dmr %in% fm_inter_dmr)
male_common_dmrs = male_dmr_assoc_res_adj %>% filter(dmr %in% fm_inter_dmr)
ai_dual_male_common_dmrs = male_common_dmrs %>% filter(cate != "age")
ai_dual_male_common_cpgs = male_fidmrs_cpgs %>% filter(dmr %in% ai_dual_male_common_dmrs$dmr)
ai_dual_male_path = prob_enriched_path(ai_dual_male_common_cpgs$IlmnID)
```

```{r}
ad_common_dmrs = setdiff(fm_inter_dmr, ai_dual_male_common_dmrs$dmr)
ad_common_cpgs = female_fidmrs_cpgs %>% filter(dmr %in% ad_common_dmrs)
ad_common_path = prob_enriched_path(ad_common_cpgs$IlmnID)
ad_common_path$term_name = factor(ad_common_path$term_name, 
                                           levels = ad_common_path$term_name)

ggplot(data = ad_common_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point()+
    scale_size(range = c(1, 3)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 8, face = "bold")) +
    scale_x_continuous(expand = c(0.05, 0.05)) +
    labs(y = "") 
```

```{r}
female_specific_dmrs = setdiff(female_fidmrs$dmr, female_common_dmrs$dmr)
female_specific_cpgs = female_fidmrs_cpgs %>% filter(dmr %in% female_specific_dmrs)
male_specific_dmrs = setdiff(male_fidmrs$dmr, female_common_dmrs$dmr)
male_specific_cpgs = male_fidmrs_cpgs %>% filter(dmr %in% male_specific_dmrs)
female_specific_path = prob_enriched_path(female_specific_cpgs$IlmnID)
```

## 3.10 female_male specific DMRs examples
```{r}
f_specific_dmr_example = female_specific_cpgs %>% filter(dmr == "chr3:61002327-61003134")
f_specific_example_tbl = cbind(methy_dat_dis_meta %>% 
                                     dplyr::select(id, mouse_id, FI.Score, Sex),
                               methy_dat_dis_M %>% 
                                     dplyr::select(f_specific_dmr_example$IlmnID)) %>%
  as.data.frame() %>% 
  melt(., id = c("id", "mouse_id", "FI.Score", "Sex")) %>%
  as.data.frame()

ggplot(f_specific_example_tbl,
       aes(x = log2(FI.Score), 
           y = value)) +
  geom_point(aes(color = Sex), size = 0.5) +
  geom_smooth(method = "lm", aes(group = Sex, color = Sex, fill = Sex), alpha = 0.2) +
  scale_color_manual(values = c("#AA4466", "#4477AA")) 
```

```{r}
m_specific_dmr_example = male_specific_cpgs %>% filter(dmr == "chr17:47596108-47597626")
m_specific_example_tbl = cbind(methy_dat_dis_meta %>% 
                                     dplyr::select(id, mouse_id, FI.Score, Sex),
                               methy_dat_dis_M %>% 
                                     dplyr::select(m_specific_dmr_example$IlmnID)) %>%
  as.data.frame() %>% 
  melt(., id = c("id", "mouse_id", "FI.Score", "Sex")) %>%
  as.data.frame()

ggplot(m_specific_example_tbl,
       aes(x = log2(FI.Score), 
           y = value)) +
  geom_point(aes(color = Sex), size = 0.5) +
  geom_smooth(method = "lm", aes(group = Sex, color = Sex, fill = Sex), alpha = 0.2) +
  scale_color_manual(values = c("#AA4466", "#4477AA")) 
```

## 3.11 female_male associations with frailty
```{r}
methy_dat_dis_meta_update = methy_dat_dis_meta %>% as.data.frame() %>%
  mutate(age = `Age.at.assesment..days.`,
         logFI = log2(FI.Score),
         sex = as.factor(Sex),
         mouse_id = as.factor(Mouse.ID)) %>%
  cbind(methy_dat_dis_M, .) %>%
  as.data.frame()
fm_all_dmr_cpgs_combined = rbind(female_fidmrs_cpgs, male_fidmrs_cpgs) %>%
  as.data.frame()
frail_dmr_assoc_mix = lapply(fm_all_dmrs, function(x) {
  dmr_cpgs = fm_all_dmr_cpgs_combined %>% filter(dmr == x)
  fm_all_dmr_cpgs_combined_melt = methy_dat_dis_meta_update %>%
    dplyr::select(unique(dmr_cpgs$IlmnID), id, mouse_id, age, sex, logFI) %>%
    melt(., id = c("id", "mouse_id", "age", "sex", "logFI")) %>%
    as.data.frame()
  dmr_inter_assoc_mix = lmer(value ~ logFI*sex + (1|mouse_id) + (0 + logFI|mouse_id), data = fm_all_dmr_cpgs_combined_melt)
  return(dmr_inter_assoc_mix)
})
```

```{r}
test_inter = lapply(1:length(fm_all_dmrs), function(x) {
  return(c(fm_all_dmrs[[x]], summary(frail_dmr_assoc_mix[[x]])$coefficient[4, 5]))
}) %>% Reduce(rbind, .) %>%
  as.data.frame() %>% 
  mutate(adjp = p.adjust(V2, method = "fdr")) %>%
           filter(adjp < 0.1)

female_sig = female_specific_dmrs[female_specific_dmrs %in% test_inter$V1]
male_sig = male_specific_dmrs[male_specific_dmrs %in% test_inter$V1]
```

## 3.10. combined fm dmrs association
```{r, message = FALSE, warning = FALSE}
fidmr_assoc_f = lapply(fm_all_dmrs, function(x) {
  dmr_cpgs = fm_all_dmr_cpgs_combined %>% filter(dmr == x)
  dmr_cpgs_melt = methy_dat_dis_meta_update %>%
    filter(sex == "female") %>%
    dplyr::select(unique(dmr_cpgs$IlmnID), id, mouse_id, age, logFI) %>%
    melt(., id = c("id", "mouse_id", "age", "logFI")) %>%
    as.data.frame()
  dmr_assoc_res = lmer(value ~ logFI + (1|mouse_id) + (0 + logFI|mouse_id), data = dmr_cpgs_melt)
  return(dmr_assoc_res)
})

fidmr_assoc_m = lapply(fm_all_dmrs, function(x) {
  dmr_cpgs = fm_all_dmr_cpgs_combined %>% filter(dmr == x)
  dmr_cpgs_melt = methy_dat_dis_meta_update %>%
    filter(sex == "male") %>%
    dplyr::select(unique(dmr_cpgs$IlmnID), id, mouse_id, age, logFI) %>%
    melt(., id = c("id", "mouse_id", "age", "logFI")) %>%
    as.data.frame()
  dmr_assoc_res = lmer(value ~ logFI + (1|mouse_id) + (0 + logFI|mouse_id), data = dmr_cpgs_melt)
  return(dmr_assoc_res)
})
```

```{r}
## comparisons of female and male associations
frail_dmr_coeff_random_fm = data.frame(dmr = fm_all_dmrs,
                                    coeff_f = sapply(1:903, function(x) 
                                      {coef = summary(fidmr_assoc_f[[x]])$coefficients[2, 1] 
                                      return(coef)}),
                                    coeff_m = sapply(1:903, function(x) 
                                      {coef = summary(fidmr_assoc_m[[x]])$coefficients[2, 1] 
                                      return(coef)})) %>%
  mutate(dmr_type = case_when(
    dmr %in% female_fidmrs$dmr & dmr %in% male_fidmrs$dmr ~ "common",
    dmr %in% female_fidmrs$dmr ~ "f",
    dmr %in% male_fidmrs$dmr ~ "m")) %>%
  mutate(dir = ifelse(coeff_f * coeff_m > 0, "1", "0")) %>%
  mutate(dmr_dir = paste(dmr_type, dir))

frail_dmr_coeff_random_fm$dmr_dir = factor(frail_dmr_coeff_random_fm$dmr_dir,
                                                    levels = c("common 1", "common 0", "f 1", "f 0", "m 1", "m 0"))
frail_dmr_coeff_random_fm$inter_sig = factor(frail_dmr_coeff_random_fm$inter_sig,
                                                    levels = c("1", "0"))

ggplot(data = frail_dmr_coeff_random_fm,
       aes(x = coeff_f,
           y = coeff_m)) +
   geom_point(aes(shape = as.factor(dmr_dir),
           color = as.factor(inter_sig)),
           size = 0.75) +
   scale_shape_manual(values = c(19, 21, 17, 24, 15, 22)) +
   scale_color_manual(values = c("#44AA99", "#762A83")) +
   coord_fixed() +
   geom_abline(slope = 1, intercept = 0, lty = 3) +
   geom_vline(xintercept = 0, lty = 3) +
   geom_hline(yintercept = 0, lty = 3) 
```

```{r}
fm_overlap = intersect(unique(female_fidmrs_cpgs$dmr), unique(male_fidmrs_cpgs$dmr))
ggplot(data = frail_dmr_coeff_random_fm %>% filter(dmr %in% fm_overlap),
       aes(x = coeff_f,
           y = coeff_m)) +
   geom_point(aes(shape = as.factor(dmr_dir)),
           #color = as.factor(inter_sig)),
           size = 0.75,
           color = "#762A83") +
   scale_shape_manual(values = c(19, 21)) +
   coord_fixed() +
   geom_abline(slope = 1, intercept = 0, lty = 3) +
   geom_vline(xintercept = 0, lty = 3) +
   geom_hline(yintercept = 0, lty = 3) 
```


```{r}
female_positive_dmrs = frail_dmr_coeff_random_fm %>% filter(dmr_type == "f" & coeff_f > 0)
female_negative_dmrs = frail_dmr_coeff_random_fm %>% filter(dmr_type == "f" & coeff_f < 0)
female_positive_dmrs_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% female_positive_dmrs$dmr)
female_positive_dmrs_path = prob_enriched_path(female_positive_dmrs_cpgs$IlmnID)
female_negative_dmrs_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% female_negative_dmrs$dmr)
female_negative_dmrs_path = prob_enriched_path(female_negative_dmrs_cpgs$IlmnID)
```

```{r}
female_positive_dmrs_path$term_name = factor(female_positive_dmrs_path$term_name, 
                                           levels = female_positive_dmrs_path$term_name)

ggplot(data = female_positive_dmrs_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point()+
    scale_size(range = c(1, 3)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 8, face = "bold")) +
    scale_x_continuous(expand = c(0.05, 0.05)) +
    labs(y = "") 
```

```{r}
female_negative_dmrs_path$term_name = factor(female_negative_dmrs_path$term_name, 
                                           levels = female_negative_dmrs_path$term_name)

ggplot(data = female_negative_dmrs_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point()+
    scale_size(range = c(1, 3)) +
    scale_color_berlin(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 8, face = "bold")) +
    scale_x_continuous(expand = c(0.05, 0.05)) +
    labs(y = "") 
```

```{r}
male_positive_dmrs = frail_dmr_coeff_random_fm %>% filter(dmr_type == "m" & coeff_f > 0)
male_negative_dmrs = frail_dmr_coeff_random_fm %>% filter(dmr_type == "m" & coeff_f < 0)
male_positive_dmrs_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% male_positive_dmrs$dmr)
male_positive_dmrs_gene = prob_enriched_genes(male_positive_dmrs_cpgs$IlmnID)
male_negative_dmrs_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% male_negative_dmrs$dmr)
male_negative_dmrs_gene = prob_enriched_genes(male_negative_dmrs_cpgs$IlmnID)
```

# 5. core dna methylation signatures

## 5.1. dmrs association with frailty 
```{r}
fidmrs_combined = Reduce(union, list(mix_fidmr$dmr,
                                       female_fidmrs$dmr,
                                       male_fidmrs$dmr))
fidmrs_cpg_combined = Reduce(rbind, list(mix_fidmr_cpgs,
                                            female_fidmrs_cpgs,
                                            male_fidmrs_cpgs)) %>%
  as.data.frame() %>%
  distinct()
mix_unique_dmr = setdiff(mix_fidmr$dmr, union(female_fidmrs$dmr, male_fidmrs$dmr))
comb_frail_cpg_assoc = methy_dat_dis_meta_update %>%
  mutate(age = as.numeric(`Age.at.assesment..days.`)/1200) %>%
  dplyr::select(unlist(fidmrs_cpg_combined$IlmnID), sex, age, `Time.label`, mouse_id, logFI, id)
```

```{r, message = FALSE, warning = FALSE}
dmr_frail_coeff = lapply(1:925, function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == unique(fidmrs_cpg_combined$dmr)[[x]])
  expr_merged = comb_frail_cpg_assoc %>%
    dplyr::select(loci_lmer$IlmnID, id, mouse_id, logFI, sex) %>%
    melt(., id = c("id", "mouse_id", "logFI", "sex")) %>%
    as.data.frame()
  res = lmer(logFI ~ value + as.factor(sex) + (1|mouse_id) + (1|variable), data = expr_merged)
  coef = summary(res)$coefficients
  return(c(coef[2, 1], coef[2, 5]))}) %>%
  Reduce(rbind, .) %>%
  as.data.frame()
```

```{r}
dmr_nocpg = table(fidmrs_cpg_combined$dmr) %>% as.data.frame()
dmr_frail_coeff_update = dmr_frail_coeff %>%
  mutate(adjp = p.adjust(V2, method = "BH")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr)) %>%
  merge(., mix_dmr_greater5, by.x = "dmr", by.y = "coord", all.x = T) %>%
  merge(., dmr_nocpg, by.x = "dmr", by.y = "Var1") %>%
  as.data.frame() %>%
  mutate(cate = case_when(
    dmr %in% fm_inter_dmr ~ "common",
    dmr %in% female_specific_dmrs ~ "f",
    dmr %in% male_specific_dmrs ~ "m",
    dmr %in% mix_unique_dmr ~ "mix"),
    MAPINFO = (start + end)/2) %>%
  merge(., chromo_cumpos, by.x = "chr", by.y = "chromo") %>%
  as.data.frame() %>%
  mutate(bpcum = MAPINFO + position_x)

ggplot() +
  geom_rect(data = chromo_cumpos, 
            aes(xmin = position_x, xmax = dplyr::lead(position_x, default = max(position_x, na.rm = TRUE)),
                ymin = -Inf, ymax = Inf), 
            fill = rep(c("#f5f0f0", "white"), 11)) +
  geom_point(data = dmr_frail_coeff_update,
             aes(x = bpcum,
                 y = -log10(adjp),
                 size = log2(Freq),
                 color = as.factor(cate),
                 shape = as.factor(cate))) +
  scale_size(range = c(1, 3.5)) +
  scale_color_manual(values = c("#DDAA33", "#AA4466", "#4477AA", "#44AA99")) +
  scale_shape_manual(values = c(15, 16, 17, 8)) + 
  scale_x_continuous(label = x_axis_ticks$chromo, breaks = x_axis_ticks$center, 
                     expand = c(0, 0)) +
  labs(x = "", y = "-log10(p)", color = "") +
  geom_hline(yintercept = 0, lty = 3, color = "grey30", size = 1) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1, size = 9, face = "bold"))
```

## 5.2. positive and negative association
```{r}
dmr_positive_dmrs = dmr_frail_coeff_update %>% filter(V1 > 0)
dmr_positive_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% dmr_positive_dmrs$dmr)
dmr_negative_dmrs = dmr_frail_coeff_update %>% filter(V1 < 0)
dmr_negative_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% dmr_negative_dmrs$dmr)
```

```{r}
dmr_pos_path = prob_enriched_path(dmr_positive_dmr_cpgs$IlmnID)
dmr_neg_path = prob_enriched_path(dmr_negative_dmr_cpgs$IlmnID)
dmr_pos_path$term_name = factor(dmr_pos_path$term_name, levels = dmr_pos_path$term_name)
dmr_neg_path$term_name = factor(dmr_neg_path$term_name, levels = dmr_neg_path$term_name)
ggplot(data = dmr_pos_path,
       aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_color_bamako(reverse = TRUE) +
    scale_size(range = c(2, 5)) +
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    scale_x_continuous(expand = c(0.1, 0.1)) +
    labs(y = "") 
```

```{r}
ggplot(data = dmr_neg_path,
       aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_color_berlin(reverse = TRUE) +
    scale_size(range = c(2, 5)) + 
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    scale_x_continuous(expand = c(0.1, 0.1)) +
    labs(y = "")
```

## 5.3. positive and negative association examples
```{r}
pos_dmr_example_cpgs = fidmrs_cpg_combined %>% filter(dmr == "chr11:57831503-57832850") %>% as.data.frame()
pos_dmr_example_tbl = cbind(methy_dat_dis_meta_update %>% 
                                     dplyr::select(id, mouse_id, logFI, Sex),
                            methy_dat_dis_M %>% 
                                     dplyr::select(pos_dmr_example_cpgs$IlmnID)) %>%
  as.data.frame() %>%
  melt(., id = c("id", "mouse_id", "logFI", "Sex")) %>%
  as.data.frame()

ggplot(pos_dmr_example_tbl,
       aes(x = logFI, 
           y = value)) +
  geom_point(aes(color = Sex), size = 0.2, alpha = 0.4) +
  geom_smooth(method = "lm", aes(group = Sex, color = Sex, fill = Sex), alpha = 0.2) +
  scale_color_manual(values = c("#AA4466", "#4477AA")) 
```

```{r}
neg_dmr_example_cpgs = fidmrs_cpg_combined %>% filter(dmr == "chr4:107537520-107538859") %>% as.data.frame()
neg_dmr_example_tbl = cbind(methy_dat_dis_meta_update %>% 
                                     dplyr::select(id, mouse_id, logFI, Sex),
                            methy_dat_dis_M %>% 
                                     dplyr::select(neg_dmr_example_cpgs$IlmnID)) %>%
  as.data.frame() %>%
  melt(., id = c("id", "mouse_id", "logFI", "Sex")) %>%
  as.data.frame()
ggplot(neg_dmr_example_tbl,
       aes(x = logFI, 
           y = value)) +
  geom_point(aes(color = Sex), size = 0.2, alpha = 0.4) +
  geom_smooth(method = "lm", aes(group = Sex, color = Sex, fill = Sex), alpha = 0.2) +
  scale_color_manual(values = c("#AA4466", "#4477AA")) 
```

# 6. DMRs association with future frailty outcomes

## 6.1 Sex-mixed analysis
```{r, message = FALSE, warning = FALSE}
## sex-mixed analysis
tp_group = data.frame(pre = c(rep("BL", 4), rep("T2", 3), rep("T3", 2), "T4"),
                      post = c("T2", "T3", "T4", "T5", "T3", "T4", "T5", "T4", "T5", "T5"))
comb_frail_cpg_assoc_prepost = prepost_data_merge(tp_group, comb_frail_cpg_assoc, fidmrs_cpg_combined$IlmnID)
tp_future_group = data.frame(pre = c(rep("BL", 6), rep("T2", 3), "T3"),
                             post = c("T2", "T2", "T2", "T3", "T3", "T4", "T3", "T3", "T4", "T4"),
                             future = c("T3", "T4", "T5", "T4", "T5", "T5", "T4", "T5", "T5", "T5"))
comb_frail_cpg_assoc_prepostfuture = prepostfuture_data_merge(tp_future_group, comb_frail_cpg_assoc, fidmrs_cpg_combined$IlmnID)

dmr_assoc_res = dmr_assoc_frailty_outcome(comb_frail_cpg_assoc, comb_frail_cpg_assoc_prepost,
                                          comb_frail_cpg_assoc_prepostfuture)

dmr_assoc_fic_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res[[x]][[1]])$coefficients
  coef2 = summary(dmr_assoc_res[[x]][[2]])$coefficients
  bound1 = confint(dmr_assoc_res[[x]][[1]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fic") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05) 

dmr_assoc_fif_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res[[x]][[3]])$coefficients
  coef2 = summary(dmr_assoc_res[[x]][[4]])$coefficients
  bound1 = confint(dmr_assoc_res[[x]][[3]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fif") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_deltafi_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res[[x]][[5]])$coefficients
  coef2 = summary(dmr_assoc_res[[x]][[6]])$coefficients
  bound1 = confint(dmr_assoc_res[[x]][[5]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_delta") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fic_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res[[x]][[7]])$coefficients
  coef2 = summary(dmr_assoc_res[[x]][[8]])$coefficients
  bound1 = confint(dmr_assoc_res[[x]][[7]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fic") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fif_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res[[x]][[9]])$coefficients
  coef2 = summary(dmr_assoc_res[[x]][[10]])$coefficients
  bound1 = confint(dmr_assoc_res[[x]][[9]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[7, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fif") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_deltafi_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res[[x]][[11]])$coefficients
  coef2 = summary(dmr_assoc_res[[x]][[12]])$coefficients
  bound1 = confint(dmr_assoc_res[[x]][[11]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[7, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_delta") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_freq = get_freq_dmr(dmr_assoc_fic_res, dmr_assoc_fif_res, 
                                     dmr_assoc_deltafi_res, deltadmr_assoc_fic_res,
                                     deltadmr_assoc_fif_res, deltadmr_assoc_deltafi_res)
dmr_assoc_freq_over2 = dmr_assoc_freq %>% filter(num >= 3) %>% arrange(desc(num))
dmr_assoc_freq_opposite = dmr_assoc_freq %>% filter(dmr %in% c("chr4:153481117-153483326", "chr2:163224441-163225596",
                                                               "chrX:141724036-141726375", "chr2:174283351-174287998",
                                                               "chr6:4746094-4748541", "chr7:142574973-142582518",
                                                               "chr10:13089935-13092324", "chr17:12741330-12742949")) %>%
  arrange(desc(dmr_fic), desc(deltadmr_fic), desc(deltadmr_fif), desc(deltadmr_deltafi))
```
```{r}
deltafi_visual = dmr_assoc_deltafi_res %>% 
  arrange(coef) %>% 
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))

deltafi_visual$dmr = factor(deltafi_visual$dmr, levels = deltafi_visual$dmr)
ggplot(data = deltafi_visual) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 3.5) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         color = "#117777")  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```


```{r}
dmr_assoc_list = Reduce(union, list(dmr_assoc_deltafi_res$dmr,
                                      deltadmr_assoc_fif_res$dmr,
                                      deltadmr_assoc_deltafi_res$dmr))
dmr_assoc_freq = data.frame(dmr_deltafi = ifelse(dmr_assoc_list %in% dmr_assoc_deltafi_res$dmr, 1, 0),
                              deltadmr_fif = ifelse(dmr_assoc_list %in% deltadmr_assoc_fif_res$dmr, 1, 0),
                              deltadmr_deltafi = ifelse(dmr_assoc_list %in% deltadmr_assoc_deltafi_res$dmr, 1, 0)) %>%
    mutate(num = rowSums(.),
           dmr = dmr_assoc_list) %>%
    filter(num >= 2) 
```

```{r}
mix_assoc_union_dmr = Reduce(union, list(dmr_assoc_deltafi_res$dmr, deltadmr_assoc_fif_res$dmr,
                                                 deltadmr_assoc_deltafi_res$dmr))

deltafi_visual
mix_assoc_union_dmr_cpgs = fidmrs_cpg_combined %>% 
  filter(dmr == "chr18:20059073-20059584")
prob_enriched_genes(mix_assoc_union_dmr_cpgs$IlmnID)

mix_assoc_greater2_cpgs = fidmrs_cpg_combined %>% 
  filter(dmr %in% dmr_assoc_freq_over2$dmr)
prob_enriched_genes(mix_assoc_greater2_cpgs$IlmnID)
```


```{r}
dmr_mix_visual_tbl_over2 = lapply(list(dmr_assoc_fif_res, dmr_assoc_deltafi_res, 
                                 deltadmr_assoc_fif_res, deltadmr_assoc_deltafi_res), function(x) {
                                   df = x %>% filter(dmr %in% dmr_assoc_freq$dmr)
                                 }) %>%
  Reduce(rbind, .) %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))
dmr_mix_visual_tbl_over2_order = merge(dmr_mix_visual_tbl_over2, dmr_assoc_freq, by = "dmr") %>% 
  as.data.frame() %>% 
  arrange(dmr_deltafi, deltadmr_fif, coef) %>% 
  dplyr::select(dmr, num) %>%
  distinct()
dmr_mix_visual_tbl_over2$dmr = factor(dmr_mix_visual_tbl_over2$dmr, 
                                      levels = dmr_mix_visual_tbl_over2_order$dmr)
dmr_mix_visual_tbl_over2$type = factor(dmr_mix_visual_tbl_over2$type, levels = c("delta_delta", "delta_fif", "dmr_delta"))

ggplot(data = dmr_mix_visual_tbl_over2) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 3.5) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type)),
                position = position_dodge(width = 0.8),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.10, 0.12) +
      scale_color_manual(values = c("#DDAA33", "#762A83", "#117777")) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```

```{r}
delta_fic_bl_adjust = lapply(c("chr4:153481117-153483326", "chrX:141724036-141726375"), function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == x)
  loci_names = paste0("delta_", loci_lmer$IlmnID)
  expr_merged = comb_frail_cpg_assoc_prepost %>%
      dplyr::select(all_of(loci_names), id.x, mouse_id, age.y, age_change, logFI.y, sex.x, logFI.x) %>%
      melt(., id = c("id.x", "mouse_id", "age.y", "age_change", "logFI.y", "sex.x", "logFI.x")) %>%
      as.data.frame()
  res = lmer(logFI.y ~ value + age.y + as.factor(sex.x) + age_change + logFI.x + (1 | mouse_id) + (1 | variable), 
                data = expr_merged)
  coef = summary(res)$coefficients
  bound = confint(res, method = "Wald")
  return(c(coef[2, 1], coef[2, 5], bound[5, 1], bound[5, 2]))
}) %>%
  Reduce(rbind, .) %>%
  `colnames<-`(c("coef", "p1", "lb", "ub")) %>%
  as.data.frame() %>%
  mutate(dmr = c("chr4:153481117-153483326", "chrX:141724036-141726375"))
```

```{r}
delta_fif_bl_adjust = lapply(c("chr2:174283351-174287998",
                               "chr6:4746094-4748541", "chr7:142574973-142582518",
                               "chr10:13089935-13092324", "chr17:12741330-12742949"), function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == x)
  loci_names = paste0("delta_", loci_lmer$IlmnID)
  expr_merged = comb_frail_cpg_assoc_prepostfuture %>%
      dplyr::select(all_of(loci_names), id.x, mouse_id, age.y, age_change1, age_change2, logFI, fi_change, sex, logFI.y) %>%
      melt(., id = c("id.x", "mouse_id", "age.y", "age_change1", "age_change2", "logFI", "fi_change", "sex", "logFI.y")) %>%
      as.data.frame()
  res = lmer(logFI ~ value + age.y + as.factor(sex) + age_change1 + age_change2 + logFI.y + (1 | mouse_id) + (1 | variable),
                data = expr_merged)
  coef = summary(res)$coefficients
  bound = confint(res, method = "Wald")
  return(c(coef[2, 1], coef[2, 5], bound[5, 1], bound[5, 2]))
}) %>%
  Reduce(rbind, .) %>%
  `colnames<-`(c("coef", "p1", "lb", "ub")) %>%
  as.data.frame() %>%
  mutate(dmr = c("chr2:174283351-174287998",
                               "chr6:4746094-4748541", "chr7:142574973-142582518",
                               "chr10:13089935-13092324", "chr17:12741330-12742949"))
```

```{r}
opposite_dmr1_cpgs = fidmrs_cpg_combined %>% filter(dmr == "chr2:174283351-174287998") %>% as.data.frame()
opposite_dmr1_cpgs_delta = paste0("delta_", opposite_dmr1_cpgs$IlmnID)
opposite_dmr1_cpgs_deltadelta_tbl = comb_frail_cpg_assoc_prepostfuture %>%
  dplyr::select(id.x, mouse_id, age.y, sex, logFI.y, logFI.x, logFI, age_change1, age_change2, fi_change, opposite_dmr1_cpgs_delta) %>%
  melt(., id = c("id.x", "mouse_id", "sex", "logFI.y", "logFI.x", "logFI", "age.y", "age_change1", "age_change2", "fi_change")) %>%
  as.data.frame()

dmr1_deltafif_y_res = lm(logFI ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), data = opposite_dmr1_cpgs_deltadelta_tbl)
dmr1_deltafif_x_res = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), data = opposite_dmr1_cpgs_deltadelta_tbl)
dmr1_deltafif_y_res_base = lm(logFI ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id) + logFI.y, data = opposite_dmr1_cpgs_deltadelta_tbl)
dmr1_deltafif_x_res_base = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + + as.factor(mouse_id) + logFI.y, data = opposite_dmr1_cpgs_deltadelta_tbl)
dmr1_deltadelta_y_res = lm(fi_change ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), 
                           data = opposite_dmr1_cpgs_deltadelta_tbl)
dmr1_deltadelta_x_res = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id),
                           data = opposite_dmr1_cpgs_deltadelta_tbl)

opposite_dmr1_cpgs_deltafif_tbl = opposite_dmr1_cpgs_deltadelta_tbl %>%
  mutate(fif_y = residuals(dmr1_deltafif_y_res),
         fif_x = residuals(dmr1_deltafif_x_res),
         fif_y_base = residuals(dmr1_deltafif_y_res_base),
         fif_x_base = residuals(dmr1_deltafif_x_res_base),
         delta_y = residuals(dmr1_deltadelta_y_res),
         delta_x = residuals(dmr1_deltadelta_x_res))

dmr1_deltafif_plot = ggplot(opposite_dmr1_cpgs_deltafif_tbl,
       aes(x = fif_y, 
           y = fif_x)) +
  geom_point(size = 0.8, color = "#762A83", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#762A83", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(log2FIf ~ covariates)")

dmr1_deltafif_plot_base = ggplot(opposite_dmr1_cpgs_deltafif_tbl,
       aes(x = fif_y_base, 
           y = fif_x_base)) +
  geom_point(size = 0.8, color = "#114477", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#114477", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates + base)",
       y = "Residual(log2FIf ~ covariates + base)")

dmr1_deltadelta_plot = ggplot(opposite_dmr1_cpgs_deltafif_tbl,
       aes(x = delta_y, 
           y = delta_x)) +
  geom_point(size = 0.8, color = "#DDAA33", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#DDAA33", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(deltaFI ~ covariates)")

ggarrange(dmr1_deltafif_plot, dmr1_deltafif_plot_base, dmr1_deltadelta_plot, ncol = 3, nrow = 1)
```

```{r}
opposite_dmr2_cpgs = fidmrs_cpg_combined %>% filter(dmr == "chrX:141724036-141726375") %>% as.data.frame()
opposite_dmr2_cpgs_delta = paste0("delta_", opposite_dmr2_cpgs$IlmnID)
opposite_dmr2_cpgs_deltafic_tbl = comb_frail_cpg_assoc_prepost %>%
  dplyr::select(id.x, mouse_id, logFI.y, sex.x, age.y, age_change, opposite_dmr1_cpgs_delta, logFI.x) %>%
  melt(., id = c("id.x", "mouse_id", "logFI.y", "sex.x", "age.y", "age_change", "logFI.x")) %>%
  as.data.frame()
dmr2_deltafic_y_res = lm(logFI.y ~ age_change + as.factor(sex.x) + age.y + as.factor(mouse_id), data = opposite_dmr2_cpgs_deltafic_tbl)
dmr2_deltafic_x_res = lm(value ~ age_change + as.factor(sex.x) + age.y + + as.factor(mouse_id), data = opposite_dmr2_cpgs_deltafic_tbl)
dmr2_deltafic_y_res_base = lm(logFI.y ~ age_change + as.factor(sex.x) + age.y + logFI.x + as.factor(mouse_id), data = opposite_dmr2_cpgs_deltafic_tbl)
dmr2_deltafic_x_res_base = lm(value ~ age_change + as.factor(sex.x) + age.y + logFI.x + as.factor(mouse_id), data = opposite_dmr2_cpgs_deltafic_tbl)
opposite_dmr2_cpgs_deltafic_tbl = opposite_dmr2_cpgs_deltafic_tbl %>%
  mutate(res_y = residuals(dmr2_deltafic_y_res),
         res_x = residuals(dmr2_deltafic_x_res),
         res_y_base = residuals(dmr2_deltafic_y_res_base),
         res_x_base = residuals(dmr2_deltafic_x_res_base))

dmr2_deltafic_plot = ggplot(opposite_dmr2_cpgs_deltafic_tbl,
       aes(x = res_x, 
           y = res_y)) +
  geom_point(size = 0.8, color = "#AA7744", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#AA7744", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(log2FIc ~ covariates)")

dmr2_deltafic_plot_base = ggplot(opposite_dmr2_cpgs_deltafic_tbl,
       aes(x = res_x_base, 
           y = res_y_base)) +
  geom_point(size = 0.8, color = "#114477", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#114477", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(log2FIc ~ covariates)")


opposite_dmr2_cpgs_deltadelta_tbl = comb_frail_cpg_assoc_prepostfuture %>%
  dplyr::select(id.x, mouse_id, age.y, sex, age_change1, age_change2, fi_change, opposite_dmr2_cpgs_delta) %>%
  melt(., id = c("id.x", "mouse_id", "sex", "age.y", "age_change1", "age_change2", "fi_change")) %>%
  as.data.frame()

dmr2_deltadelta_y_res = lm(fi_change ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), 
                           data = opposite_dmr2_cpgs_deltadelta_tbl)
dmr2_deltadelta_x_res = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id),
                           data = opposite_dmr2_cpgs_deltadelta_tbl)
opposite_dmr2_cpgs_deltadelta_tbl = opposite_dmr2_cpgs_deltadelta_tbl %>%
  mutate(res_y = residuals(dmr2_deltadelta_y_res),
         res_x = residuals(dmr2_deltadelta_x_res))

dmr2_deltadelta_plot = ggplot(opposite_dmr2_cpgs_deltadelta_tbl,
       aes(x = res_x, 
           y = res_y)) +
  geom_point(size = 0.8, color = "#88CC88", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#88CC88", alpha = 0.2) +
  labs(x = "Residual(M-value ~ covariates)",
       y = "Residual(log2FIc ~ covariates)")
ggarrange(dmr2_deltafic_plot, dmr2_deltafic_plot_base, dmr2_deltadelta_plot, ncol = 3, nrow = 1)
```

```{r}
opposite_dmr3_cpgs = fidmrs_cpg_combined %>% filter(dmr == "chr7:142574973-142582518") %>% as.data.frame()
opposite_dmr3_cpgs_delta = paste0("delta_", opposite_dmr3_cpgs$IlmnID)
opposite_dmr3_cpgs_delta_tbl = comb_frail_cpg_assoc_prepostfuture %>%
  dplyr::select(id.x, mouse_id, age.y, sex, logFI, age_change1, age_change2, fi_change, opposite_dmr3_cpgs_delta, logFI.y) %>%
  melt(., id = c("id.x", "mouse_id", "sex", "age.y", "age_change1", "age_change2", "fi_change", "logFI", "logFI.y")) %>%
  as.data.frame()

dmr3_deltafif_y_res = lm(logFI ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), 
                           data = opposite_dmr3_cpgs_delta_tbl)
dmr3_deltafif_x_res = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id),
                           data = opposite_dmr3_cpgs_delta_tbl)

dmr3_deltafif_y_res_base = lm(logFI ~ age_change1 + age_change2 + as.factor(sex) + age.y + logFI.y + as.factor(mouse_id), 
                           data = opposite_dmr3_cpgs_delta_tbl)
dmr3_deltafif_x_res_base = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + logFI.y + as.factor(mouse_id),
                           data = opposite_dmr3_cpgs_delta_tbl)

dmr3_deltadelta_y_res = lm(fi_change ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), 
                           data = opposite_dmr3_cpgs_delta_tbl)
dmr3_deltadelta_x_res = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id),
                           data = opposite_dmr3_cpgs_delta_tbl)

opposite_dmr3_cpgs_delta_tbl = opposite_dmr3_cpgs_delta_tbl %>%
  mutate(res_y_fif = residuals(dmr3_deltafif_y_res),
         res_x_fif = residuals(dmr3_deltafif_x_res),
         res_y_fif_base = residuals(dmr3_deltafif_y_res_base),
         res_x_fif_base = residuals(dmr3_deltafif_x_res_base),
         res_y_delta = residuals(dmr3_deltadelta_y_res),
         res_x_delta = residuals(dmr3_deltadelta_x_res))

dmr3_deltafif_plot = ggplot(opposite_dmr3_cpgs_delta_tbl,
       aes(x = res_x_fif, 
           y = res_y_fif)) +
  geom_point(size = 0.8, color = "#DDAA33", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#DDAA33", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(log2FIf ~ covariates)")


dmr3_deltafif_plot_base = ggplot(opposite_dmr3_cpgs_delta_tbl,
       aes(x = res_x_fif_base, 
           y = res_y_fif_base)) +
  geom_point(size = 0.8, color = "#114477", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#114477", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates + base)",
       y = "Residual(log2FIf ~ covariates + base)")

dmr3_deltadelta_plot = ggplot(opposite_dmr3_cpgs_delta_tbl,
       aes(x = res_x_delta, 
           y = res_y_delta)) +
  geom_point(size = 0.8, color = "#88CC88", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#88CC88", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(deltaFI ~ covariates)")

ggarrange(dmr3_deltafif_plot, dmr3_deltafif_plot_base, dmr3_deltadelta_plot, ncol = 3, nrow = 1)
```

```{r}
dmr_mix_visual_tbl_opposite = lapply(list(dmr_assoc_fic_res, dmr_assoc_fif_res, dmr_assoc_deltafi_res, 
                                 deltadmr_assoc_fic_res, deltadmr_assoc_fif_res, deltadmr_assoc_deltafi_res), function(x) {
                                   df = x %>% filter(dmr %in% dmr_assoc_freq_opposite$dmr)
                                 }) %>%
  Reduce(rbind, .) %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type) %>%
  as.data.frame()

dmr_mix_visual_tbl_opposite_base = rbind(delta_fic_bl_adjust, delta_fif_bl_adjust) %>%
  as.data.frame() %>%
  mutate(adjp1 = p1,
         type = "adjust") %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type)

dmr_mix_visual_tbl_opposite_added_base = rbind(dmr_mix_visual_tbl_opposite, dmr_mix_visual_tbl_opposite_base) %>%
  as.data.frame() %>%
  mutate(adjp1 = as.numeric(adjp1)) %>%
  mutate(sig = case_when(adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))
dmr_mix_visual_tbl_opposite_added_base$dmr = factor(dmr_mix_visual_tbl_opposite_added_base$dmr, 
                                                    levels = c("chr2:174283351-174287998", "chr6:4746094-4748541",
                                                               "chr17:12741330-12742949", "chr7:142574973-142582518",
                                                               "chr10:13089935-13092324", "chrX:141724036-141726375",
                                                               "chr2:163224441-163225596", "chr4:153481117-153483326"))

dmr_mix_visual_tbl_opposite_added_base$type = factor(dmr_mix_visual_tbl_opposite_added_base$type, levels = c("adjust", "delta_delta", "delta_fif", 
                                                                                 "delta_fic", "dmr_fic"))
ggplot(data = dmr_mix_visual_tbl_opposite_added_base) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 6) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type)),
                position = position_dodge(width = 0.8),
                hjust = 1.2, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(c(-0.12, 0.15)) +
      scale_color_manual(values = c("#114477", "#88CC88", "#DDAA33", "#AA7744", "#762A83")) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```

```{r}
comb_frail_cpg_assoc_female = comb_frail_cpg_assoc %>% filter(sex == "female")
comb_frail_cpg_assoc_male = comb_frail_cpg_assoc %>% filter(sex == "male")
comb_frail_cpg_assoc_prepost_female = comb_frail_cpg_assoc_prepost %>% filter(sex.x == "female")
comb_frail_cpg_assoc_prepost_male = comb_frail_cpg_assoc_prepost %>% filter(sex.x == "male")
comb_frail_cpg_assoc_prepostfuture_female = comb_frail_cpg_assoc_prepostfuture %>% filter(sex == "female")
comb_frail_cpg_assoc_prepostfuture_male = comb_frail_cpg_assoc_prepostfuture %>% filter(sex == "male")
```

## 6.2 female-specific analysis in association
```{r, warning = FALSE, message = FALSE}
dmr_assoc_res_female = dmr_assoc_frailty_outcome_sex(comb_frail_cpg_assoc_female, comb_frail_cpg_assoc_prepost_female,
                                          comb_frail_cpg_assoc_prepostfuture_female)

dmr_assoc_fic_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_female[[x]][[1]])$coefficients
  coef2 = summary(dmr_assoc_res_female[[x]][[2]])$coefficients
  bound1 = confint(dmr_assoc_res_female[[x]][[1]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[4, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fic") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_fif_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_female[[x]][[3]])$coefficients
  coef2 = summary(dmr_assoc_res_female[[x]][[4]])$coefficients
  bound1 = confint(dmr_assoc_res_female[[x]][[3]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fif") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_deltafi_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_female[[x]][[5]])$coefficients
  coef2 = summary(dmr_assoc_res_female[[x]][[6]])$coefficients
  bound1 = confint(dmr_assoc_res_female[[x]][[5]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_delta") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fic_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_female[[x]][[7]])$coefficients
  coef2 = summary(dmr_assoc_res_female[[x]][[8]])$coefficients
  bound1 = confint(dmr_assoc_res_female[[x]][[7]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fic") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fif_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_female[[x]][[9]])$coefficients
  coef2 = summary(dmr_assoc_res_female[[x]][[10]])$coefficients
  bound1 = confint(dmr_assoc_res_female[[x]][[9]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fif") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_deltafi_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_female[[x]][[11]])$coefficients
  coef2 = summary(dmr_assoc_res_female[[x]][[12]])$coefficients
  bound1 = confint(dmr_assoc_res_female[[x]][[11]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_delta") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_freq_female = get_freq_dmr(dmr_assoc_fic_res_female, dmr_assoc_fif_res_female, 
                                     dmr_assoc_deltafi_res_female, deltadmr_assoc_fic_res_female,
                                     deltadmr_assoc_fif_res_female, deltadmr_assoc_deltafi_res_female)
```

```{r}
dmr_assoc_list = Reduce(union, list(dmr_assoc_deltafi_res_female$dmr,
                                      deltadmr_assoc_fif_res_female$dmr,
                                      deltadmr_assoc_deltafi_res_female$dmr))
dmr_assoc_freq = data.frame(dmr_deltafi = ifelse(dmr_assoc_list %in% dmr_assoc_deltafi_res_female$dmr, 1, 0),
                              deltadmr_fif = ifelse(dmr_assoc_list %in% deltadmr_assoc_fif_res_female$dmr, 1, 0),
                              deltadmr_deltafi = ifelse(dmr_assoc_list %in% deltadmr_assoc_deltafi_res_female$dmr, 1, 0)) %>%
    mutate(num = rowSums(.),
           dmr = dmr_assoc_list) %>%
    filter(num >= 2) 
```

```{r}
dmr_female_visual_tbl_over2 = lapply(list(dmr_assoc_fif_res_female, dmr_assoc_deltafi_res_female, 
                                 deltadmr_assoc_fif_res_female, deltadmr_assoc_deltafi_res_female), function(x) {
                                   df = x %>% filter(dmr %in% dmr_assoc_freq$dmr)
                                 }) %>%
  Reduce(rbind, .) %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))
dmr_female_visual_tbl_over2_order = merge(dmr_female_visual_tbl_over2, dmr_assoc_freq, by = "dmr") %>% 
  as.data.frame() %>% 
  arrange(dmr_deltafi, deltadmr_fif, coef) %>% 
  dplyr::select(dmr, num) %>%
  distinct()
dmr_female_visual_tbl_over2$dmr = factor(dmr_female_visual_tbl_over2$dmr, 
                                      levels = dmr_female_visual_tbl_over2_order$dmr)
dmr_female_visual_tbl_over2$type = factor(dmr_female_visual_tbl_over2$type, levels = c("delta_delta", "delta_fif", "dmr_delta"))

ggplot(data = dmr_female_visual_tbl_over2) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 4.5) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type)),
                position = position_dodge(width = 0.8),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.07, 0.13) +
      scale_color_manual(values = c("#DDAA33", "#762A83", "#117777")) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```

```{r}
female_assoc_union_dmr = Reduce(union, list(dmr_assoc_deltafi_res_female$dmr, 
                                            deltadmr_assoc_fif_res_female$dmr, deltadmr_assoc_deltafi_res_female$dmr))
female_assoc_union_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% female_assoc_union_dmr)
female_assoc_union_path = prob_enriched_path(female_assoc_union_dmr_cpgs$IlmnID)

female_assoc_union_path$term_name = factor(female_assoc_union_path$term_name, female_assoc_union_path$term_name)
ggplot(data = female_assoc_union_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_size(range = c(3, 5)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    scale_x_continuous(expand = c(0.07, 0.07)) +
    labs(y = "")
```

```{r}
freq_2 = dmr_assoc_freq_female %>% filter(num >=2) %>% arrange(desc(num))
freq_2_coef = lapply(freq_2$dmr, function(x){
  coef_list = sapply(list(dmr_assoc_fic_res_female, dmr_assoc_fif_res_female, 
              dmr_assoc_deltafi_res_female, deltadmr_assoc_fic_res_female,
              deltadmr_assoc_fif_res_female, deltadmr_assoc_deltafi_res_female), function(y) {
                if (!x %in% y$dmr) {
                  coef = NA
                } else {
                  coef = y[y$dmr == x, ]$coef
                }
              return(coef)
              })
  return(coef_list)
}) %>%
  Reduce(rbind, .) %>%
  as.data.frame() %>%
  mutate(dmr = freq_2$dmr)
```

```{r}
opposite_dmr_val_cpgs = fidmrs_cpg_combined %>% filter(dmr == "chr10:127664929-127667317") %>% as.data.frame()
opposite_dmr_val_cpgs_delta = paste0("delta_", opposite_dmr_val_cpgs$IlmnID)

opposite_dmr_dis_deltadelta_tbl = comb_frail_cpg_assoc_prepostfuture %>%
  dplyr::select(id.x, mouse_id, age.y, sex, logFI.y, logFI.x, logFI, age_change1, age_change2, fi_change, opposite_dmr_val_cpgs_delta) %>%
  melt(., id = c("id.x", "mouse_id", "sex", "logFI.y", "logFI.x", "logFI", "age.y", "age_change1", "age_change2", "fi_change")) %>%
  as.data.frame()
opposite_dmr_val_deltadelta_tbl = comb_frail_cpg_assoc_prepostfuture_val %>%
  dplyr::select(id.x, mouse_id, age.y, sex, logFI.y, logFI.x, logFI, age_change1, age_change2, fi_change, opposite_dmr_val_cpgs_delta) %>%
  melt(., id = c("id.x", "mouse_id", "sex", "logFI.y", "logFI.x", "logFI", "age.y", "age_change1", "age_change2", "fi_change")) %>%
  as.data.frame()

dmr_deltadelta_y_res_dis = lm(fi_change ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), 
                           data = opposite_dmr_dis_deltadelta_tbl)
dmr_deltadelta_x_res_dis = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id),
                           data = opposite_dmr_dis_deltadelta_tbl)
dmr_deltadelta_y_res_val = lm(fi_change ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id), 
                           data = opposite_dmr_val_deltadelta_tbl)
dmr_deltadelta_x_res_val = lm(value ~ age_change1 + age_change2 + as.factor(sex) + age.y + as.factor(mouse_id),
                           data = opposite_dmr_val_deltadelta_tbl)

opposite_dmr_cpgs_deltadelta_dis = opposite_dmr_dis_deltadelta_tbl %>%
  mutate(y = residuals(dmr_deltadelta_y_res_dis),
         x = residuals(dmr_deltadelta_x_res_dis))
opposite_dmr_cpgs_deltadelta_val = opposite_dmr_val_deltadelta_tbl %>%
  mutate(y = residuals(dmr_deltadelta_y_res_val),
         x = residuals(dmr_deltadelta_x_res_val))

opposite_dmr_dis_plot = ggplot(opposite_dmr_cpgs_deltadelta_dis,
       aes(x = y, 
           y = x)) +
  geom_point(size = 0.8, color = "#DDAA33", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#DDAA33", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(deltaFI ~ covariates)")

opposite_dmr_val_plot = ggplot(opposite_dmr_cpgs_deltadelta_val,
       aes(x = y, 
           y = x)) +
  geom_point(size = 0.8, color = "#DDAA33", alpha = 0.3) +
  geom_smooth(aes(group = 1), formula = 'y ~ x', method = "lm", color = "#DDAA33", alpha = 0.2) +
  labs(x = "Residual(deltaM ~ covariates)",
       y = "Residual(deltaFI ~ covariates)")

ggarrange(opposite_dmr_dis_plot, opposite_dmr_val_plot,
          ncol = 2, nrow = 1)
```

```{r}
female_opposite_dmrs = c("chr2:174283351-174287998", "chr2:174294260-174301062", "chr6:30734932-30742754",
                         "chr7:6726712-6731818", "chr10:13089935-13092324", "chr11:22971841-22974382",
                         "chr17:12741330-12742949", "chr19:58453709-58456074")
female_dmr_visual_tbl_opposite = lapply(list(dmr_assoc_deltafi_res_female, deltadmr_assoc_fic_res_female, 
                                    deltadmr_assoc_fif_res_female, deltadmr_assoc_deltafi_res_female), function(x) {
                                   df = x %>% filter(dmr %in% female_opposite_dmrs)
                                   return(df)}) %>%
  Reduce(rbind, .) %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type) %>%
  as.data.frame()

female_delta_fic_bl_adjust = lapply(c("chr2:174294260-174301062", "chr6:30734932-30742754",
                         "chr7:6726712-6731818", "chr10:13089935-13092324", "chr11:22971841-22974382",
                         "chr17:12741330-12742949"), function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == x)
  loci_names = paste0("delta_", loci_lmer$IlmnID)
  expr_merged = comb_frail_cpg_assoc_prepost_female %>%
      dplyr::select(all_of(loci_names), id.x, mouse_id, age.y, age_change, logFI.y, logFI.x) %>%
      melt(., id = c("id.x", "mouse_id", "age.y", "age_change", "logFI.y", "logFI.x")) %>%
      as.data.frame()
  res = lmer(logFI.y ~ value + age.y + age_change + logFI.x + (1 | mouse_id) + (1 | variable), 
                data = expr_merged)
  coef = summary(res)$coefficients
  bound = confint(res, method = "Wald")
  return(c(coef[2, 1], coef[2, 5], bound[5, 1], bound[5, 2]))
}) %>%
  Reduce(rbind, .) %>%
  `colnames<-`(c("coef", "p1", "lb", "ub")) %>%
  as.data.frame() %>%
  mutate(dmr = c("chr2:174294260-174301062", "chr6:30734932-30742754",
                         "chr7:6726712-6731818", "chr10:13089935-13092324", "chr11:22971841-22974382",
                         "chr17:12741330-12742949"))

female_delta_fif_bl_adjust = lapply(c("chr2:174283351-174287998", "chr19:58453709-58456074"), function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == x)
  loci_names = paste0("delta_", loci_lmer$IlmnID)
  expr_merged = comb_frail_cpg_assoc_prepostfuture_female %>%
      dplyr::select(all_of(loci_names), id.x, mouse_id, age.y, age_change1, age_change2, logFI, fi_change, logFI.y) %>%
      melt(., id = c("id.x", "mouse_id", "age.y", "age_change1", "age_change2", "logFI", "fi_change", "logFI.y")) %>%
      as.data.frame()
  res = lmer(logFI ~ value + age.y + age_change1 + age_change2 + logFI.y + (1 | mouse_id) + (1 | variable),
                data = expr_merged)
  coef = summary(res)$coefficients
  bound = confint(res, method = "Wald")
  return(c(coef[2, 1], coef[2, 5], bound[5, 1], bound[5, 2]))
}) %>%
  Reduce(rbind, .) %>%
  `colnames<-`(c("coef", "p1", "lb", "ub")) %>%
  as.data.frame() %>%
  mutate(dmr = c("chr2:174283351-174287998", "chr19:58453709-58456074"))
```

```{r}
female_dmr_mix_visual_tbl_opposite_base = rbind(female_delta_fic_bl_adjust, female_delta_fif_bl_adjust) %>%
  as.data.frame() %>%
  mutate(adjp1 = p1,
         type = "adjust") %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type)

female_dmr_mix_visual_tbl_opposite_added_base = rbind(female_dmr_visual_tbl_opposite, female_dmr_mix_visual_tbl_opposite_base) %>%
  as.data.frame() %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))
female_dmr_mix_visual_tbl_opposite_added_base$dmr = factor(female_dmr_mix_visual_tbl_opposite_added_base$dmr, 
                                                    levels = c("chr19:58453709-58456074", "chr17:12741330-12742949", 
                                                               "chr7:6726712-6731818", "chr6:30734932-30742754", 
                                                               "chr11:22971841-22974382", "chr10:13089935-13092324", 
                                                               "chr2:174294260-174301062", "chr2:174283351-174287998"))
female_dmr_mix_visual_tbl_opposite_added_base$type = factor(female_dmr_mix_visual_tbl_opposite_added_base$type, 
                                                            levels = c("adjust", "delta_delta", "delta_fif", 
                                                                                 "delta_fic", "dmr_delta"))

ggplot(data = female_dmr_mix_visual_tbl_opposite_added_base) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 7) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type)),
                position = position_dodge(width = 0.8),
                hjust = 1.2, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.08, 0.15) +
      scale_color_manual(values = c("#114477", "#88CC88", "#DDAA33", "#AA7744", "#117777")) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```

## 6.3 male-specific analysis in association
```{r, warning = FALSE, message = FALSE}
dmr_assoc_res_male = dmr_assoc_frailty_outcome_sex(comb_frail_cpg_assoc_male, comb_frail_cpg_assoc_prepost_male,
                                          comb_frail_cpg_assoc_prepostfuture_male)

dmr_assoc_fic_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_male[[x]][[1]])$coefficients
  coef2 = summary(dmr_assoc_res_male[[x]][[2]])$coefficients
  bound1 = confint(dmr_assoc_res_male[[x]][[1]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[4, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fic") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_fif_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_male[[x]][[3]])$coefficients
  coef2 = summary(dmr_assoc_res_male[[x]][[4]])$coefficients
  bound1 = confint(dmr_assoc_res_male[[x]][[3]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fif") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_deltafi_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_male[[x]][[5]])$coefficients
  coef2 = summary(dmr_assoc_res_male[[x]][[6]])$coefficients
  bound1 = confint(dmr_assoc_res_male[[x]][[5]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_delta") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fic_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_male[[x]][[7]])$coefficients
  coef2 = summary(dmr_assoc_res_male[[x]][[8]])$coefficients
  bound1 = confint(dmr_assoc_res_male[[x]][[7]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fic") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fif_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_male[[x]][[9]])$coefficients
  coef2 = summary(dmr_assoc_res_male[[x]][[10]])$coefficients
  bound1 = confint(dmr_assoc_res_male[[x]][[9]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fif") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_deltafi_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_male[[x]][[11]])$coefficients
  coef2 = summary(dmr_assoc_res_male[[x]][[12]])$coefficients
  bound1 = confint(dmr_assoc_res_male[[x]][[11]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_delta") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_freq_male = get_freq_dmr(dmr_assoc_fic_res_male, dmr_assoc_fif_res_male, 
                                     dmr_assoc_deltafi_res_male, deltadmr_assoc_fic_res_male,
                                     deltadmr_assoc_fif_res_male, deltadmr_assoc_deltafi_res_male)
```

```{r}
dmr_assoc_list = Reduce(union, list( dmr_assoc_fif_res_male$dmr,
                                      dmr_assoc_deltafi_res_male$dmr,
                                      deltadmr_assoc_fif_res_male$dmr,
                                      deltadmr_assoc_deltafi_res_male$dmr))
dmr_assoc_freq = data.frame(  dmr_fif = ifelse(dmr_assoc_list %in% dmr_assoc_fif_res_male$dmr, 1, 0),
                              dmr_deltafi = ifelse(dmr_assoc_list %in% dmr_assoc_deltafi_res_male$dmr, 1, 0),
                              deltadmr_fif = ifelse(dmr_assoc_list %in% deltadmr_assoc_fif_res_male$dmr, 1, 0),
                              deltadmr_deltafi = ifelse(dmr_assoc_list %in% deltadmr_assoc_deltafi_res_male$dmr, 1, 0)) %>%
    mutate(num = rowSums(.),
           dmr = dmr_assoc_list) %>%
    filter(num >= 2)
```

```{r}
male_assoc_union_dmr = Reduce(union, list(dmr_assoc_fif_res_male$dmr, dmr_assoc_deltafi_res_male$dmr,
                                          deltadmr_assoc_fif_res_male$dmr,      
                                          deltadmr_assoc_deltafi_res_male$dmr))
male_assoc_union_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% male_assoc_union_dmr)
male_assoc_union_path = prob_enriched_path(male_assoc_union_dmr_cpgs$IlmnID)
```

```{r}
fm_inter_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% intersect(female_assoc_union_dmr, male_assoc_union_dmr))
fm_inter_dmr_path = prob_enriched_path(fm_inter_dmr_cpgs$IlmnID)
```

```{r}
mix_fm_comb_union_dmr = Reduce(union, list(mix_assoc_union_dmr,
                                           female_assoc_union_dmr,
                                           male_assoc_union_dmr))
mix_fm_comb_union_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% mix_fm_comb_union_dmr)
mix_fm_comb_union_dmr_path = prob_enriched_path(mix_fm_comb_union_dmr_cpgs$IlmnID)

mix_fm_comb_union_dmr_path$term_name = factor(mix_fm_comb_union_dmr_path$term_name, mix_fm_comb_union_dmr_path$term_name)

ggplot(data = mix_fm_comb_union_dmr_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_size(range = c(3, 5)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    scale_x_continuous(expand = c(0.07, 0.07)) +
    labs(y = "")
```

```{r}
great1_comb_union_dmr = Reduce(union, list(mix_assoc_greater1_cpgs$dmr,
                                           female_assoc_greater1_cpgs$dmr,
                                           male_assoc_greater1_cpgs$dmr))
great1_comb_union_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% great1_comb_union_dmr)
great1_comb_union_dmr_path = prob_enriched_path(great1_comb_union_dmr_cpgs$IlmnID)
```

```{r}
freq_2 = dmr_assoc_freq_male %>% filter(num >=2) %>% arrange(desc(num))
freq_2_coef = lapply(freq_2$dmr, function(x){
  coef_list = sapply(list(dmr_assoc_fic_res_male, dmr_assoc_fif_res_male, 
              dmr_assoc_deltafi_res_male, deltadmr_assoc_fic_res_male,
              deltadmr_assoc_fif_res_male, deltadmr_assoc_deltafi_res_male), function(y) {
                if (!x %in% y$dmr) {
                  coef = NA
                } else {
                  coef = y[y$dmr == x, ]$coef
                }
              return(coef)
              })
  return(coef_list)
}) %>%
  Reduce(rbind, .) %>%
  as.data.frame() %>%
  mutate(dmr = freq_2$dmr)
```

```{r}
male_opposite_dmrs = c("chr16:16526867-16527631", "chr17:12741330-12742949")
male_dmr_visual_tbl_opposite = lapply(list(dmr_assoc_fic_res_male, dmr_assoc_fif_res_male,
                                  dmr_assoc_deltafi_res_male, deltadmr_assoc_fic_res_male, 
                                  deltadmr_assoc_fif_res_male, deltadmr_assoc_deltafi_res_male), function(x) {
                                   df = x %>% filter(dmr %in% male_opposite_dmrs)
                                   return(df)}) %>%
  Reduce(rbind, .) %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type) %>%
  as.data.frame()

male_delta_fif_bl_adjust = lapply(c("chr16:16526867-16527631", "chr17:12741330-12742949"), function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == x)
  loci_names = paste0("delta_", loci_lmer$IlmnID)
  expr_merged = comb_frail_cpg_assoc_prepostfuture_male %>%
      dplyr::select(all_of(loci_names), id.x, mouse_id, age.y, age_change1, age_change2, logFI, fi_change, logFI.y) %>%
      melt(., id = c("id.x", "mouse_id", "age.y", "age_change1", "age_change2", "logFI", "fi_change", "logFI.y")) %>%
      as.data.frame()
  res = lmer(logFI ~ value + age.y + age_change1 + age_change2 + logFI.y + (1 | mouse_id) + (1 | variable),
                data = expr_merged)
  coef = summary(res)$coefficients
  bound = confint(res, method = "Wald")
  return(c(coef[2, 1], coef[2, 5], bound[5, 1], bound[5, 2]))
}) %>%
  Reduce(rbind, .) %>%
  `colnames<-`(c("coef", "p1", "lb", "ub")) %>%
  as.data.frame() %>%
  mutate(dmr = c("chr16:16526867-16527631", "chr17:12741330-12742949"))
```

```{r}
male_dmr_mix_visual_tbl_opposite_base = male_delta_fif_bl_adjust %>%
  as.data.frame() %>%
  mutate(adjp1 = p1,
         type = "adjust") %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type)

male_dmr_mix_visual_tbl_opposite_added_base = rbind(male_dmr_visual_tbl_opposite, male_dmr_mix_visual_tbl_opposite_base) %>%
  as.data.frame() %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))
male_dmr_mix_visual_tbl_opposite_added_base$dmr = factor(male_dmr_mix_visual_tbl_opposite_added_base$dmr, 
                                                    levels = male_opposite_dmrs)

male_dmr_mix_visual_tbl_opposite_added_base$type = factor(male_dmr_mix_visual_tbl_opposite_added_base$type, 
                                                            levels = c("adjust", "delta_delta", "delta_fif", 
                                                                                 "delta_fic", "dmr_fic"))
ggplot(data = male_dmr_mix_visual_tbl_opposite_added_base) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 7) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type)),
                position = position_dodge(width = 0.8),
                hjust = 1.2, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.18, 0.2) +
      scale_color_manual(values = c("#114477", "#88CC88", "#DDAA33", "#AA7744", "#762A83")) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```

```{r}
dmr_male_visual_tbl_over2 = lapply(list(dmr_assoc_fif_res_male, dmr_assoc_deltafi_res_male, 
                                 deltadmr_assoc_fif_res_male, deltadmr_assoc_deltafi_res_male), function(x) {
                                   df = x %>% filter(dmr %in% dmr_assoc_freq$dmr)
                                 }) %>%
  Reduce(rbind, .) %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ ""))
dmr_male_visual_tbl_over2$type = factor(dmr_male_visual_tbl_over2$type, levels = c("delta_delta", "delta_fif"))
dmr_male_visual_tbl_over2$dmr = factor(dmr_male_visual_tbl_over2$dmr, levels = c("chr17:12741330-12742949", "chr16:16526867-16527631"))

ggplot(data = dmr_male_visual_tbl_over2) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 3.5) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type)),
                position = position_dodge(width = 0.8),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.15, 0.13) +
      scale_color_manual(values = c("#DDAA33", "#762A83")) +
      theme(axis.text.y = element_text(size = 8, face = "bold")) 
```

## 6.4 combination of candidate DMRs 
```{r}
mix_sex_future_dmrs = union(dmr_assoc_deltafi_res$dmr, deltadmr_assoc_deltafi_res$dmr)
female_future_dmrs = union(dmr_assoc_deltafi_res_female$dmr, deltadmr_assoc_deltafi_res_female$dmr)
male_future_dmrs = union(dmr_assoc_deltafi_res_male$dmr, deltadmr_assoc_deltafi_res_male$dmr)
all_future_dmrs = Reduce(union, list(mix_sex_future_dmrs,
                                     female_future_dmrs, 
                                     male_future_dmrs))
all_future_dmrs_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% all_future_dmrs)
all_future_dmrs_path = prob_enriched_path(all_future_dmrs_cpgs$IlmnID)

all_future_dmrs_path$term_name = factor(all_future_dmrs_path$term_name, all_future_dmrs_path$term_name)
pdf(file = "~/Desktop/DNA methylation/figures_0718/fig5f.pdf", width = 5.5, height = 2.3)
ggplot(data = all_future_dmrs_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_size(range = c(3, 5)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    scale_x_continuous(expand = c(0.05, 0.05)) +
    labs(y = "")
```

```{r}
all_greater2_dmrs = Reduce(union, list(dmr_assoc_freq$dmr,
                                       dmr_assoc_freq_female$dmr,
                                       dmr_assoc_freq_male$dmr))

all_greater2_dmrs_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% all_greater2_dmrs)
all_greater2_dmrs_genes = prob_enriched_genes(all_greater2_dmrs_cpgs$IlmnID)

all_greater2_dmrs_path$term_name = factor(all_greater2_dmrs_path$term_name, all_greater2_dmrs_path$term_name)
ggplot(data = all_greater2_dmrs_path,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_size(range = c(3, 5)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    scale_x_continuous(expand = c(0.05, 0.05)) +
    labs(y = "")
```

```{r}
opposite_dmrs_all = c("chr4:153481117-153483326", "chr2:163224441-163225596",
                      "chrX:141724036-141726375", "chr2:174283351-174287998",
                      "chr6:4746094-4748541", "chr7:142574973-142582518",
                      "chr10:13089935-13092324", "chr17:12741330-12742949",
                      "chr19:58453709-58456074", "chr17:12741330-12742949", 
                      "chr7:6726712-6731818", "chr6:30734932-30742754", 
                      "chr11:22971841-22974382", "chr10:13089935-13092324", 
                      "chr2:174294260-174301062", "chr2:174283351-174287998",
                      "chr16:16526867-16527631", "chr17:12741330-12742949") %>% unique()
opposite_dmrs_all_cpgs = lapply(opposite_dmrs_all, function(x) {
                                    df = fidmrs_cpg_combined %>% filter(dmr == x)
                                    return(unlist(df$IlmnID))
                                  }) %>% Reduce(c, .)

opposite_dmrs_cpgs_genes = prob_enriched_genes(opposite_dmrs_all_cpgs)
opposite_dmrs_cpgs_paths = prob_enriched_path(opposite_dmrs_all_cpgs)

opposite_dmrs_cpgs_paths$term_name = factor(opposite_dmrs_cpgs_paths$term_name, opposite_dmrs_cpgs_paths$term_name)
ggplot(data = opposite_dmrs_cpgs_paths,
                aes(x = intersection_size,
                    y = term_name,
                    color = -log10(p_value),
                    size = prop)) +
    geom_point() +
    scale_size(range = c(3, 5)) +
    scale_color_bamako(reverse = TRUE) +
    theme(axis.text.y = element_text(size = 9, face = "bold")) +
    labs(y = "")
```

## 6.5 validation cohort dmr association, sex-mixed 
```{r}
methy_dat_val_comb_frail_cpg = methy_dat_val %>% ## need imputate
  dplyr::select(unlist(fidmrs_cpg_combined$IlmnID)) %>%
  mutate(id = rownames(.)) %>%
  merge(., methy_dat_val_meta, by = "id") %>%
  mutate(sex_tp = paste(Sex, Time.label)) 

methy_dat_val_impute = lapply(unique(methy_dat_val_comb_frail_cpg$sex_tp), function(x) {
  methy_dat_val_sextp = methy_dat_val_comb_frail_cpg %>%
    filter(sex_tp == x) 
  methy_dat_val_sextp_impute = methy_dat_val_sextp %>%
    dplyr::select(unlist(fidmrs_cpg_combined$IlmnID)) %>%
    base::apply(., 2, function(y) replace(as.numeric(y), is.na(y), mean(as.numeric(y), na.rm = TRUE))) %>%
      as.data.frame() %>%
    mutate(id = methy_dat_val_sextp$id)
  return(methy_dat_val_sextp_impute)
}) %>%
  Reduce(rbind, .) %>%
  as.data.frame() %>%
  `rownames<-`(.$id) %>%
  dplyr::select(-id)

methy_dat_val_M = methy_dat_val_impute %>% apply(., 2, beta2m) %>% as.data.frame()

comb_frail_cpg_assoc_val = methy_dat_val_M %>%
  dplyr::select(all_of(fidmrs_cpg_combined$IlmnID)) %>%
  mutate(id = rownames(.)) %>%
  merge(.,  methy_dat_val_meta, by = "id") %>%
                  as.data.frame() %>%
  mutate(age = as.numeric(`Age.at.assesment..days.`)/1200,
         logFI = log2(FI.Score),
         sex = ifelse(Sex == "female", 1, 0),
         mouse_id = as.factor(`Mouse.ID`)) %>%
  dplyr::select(unlist(fidmrs_cpg_combined$IlmnID), age, sex, `Time.label`, 
                mouse_id, logFI, id)

comb_frail_cpg_assoc_prepost_val = prepost_data_merge(tp_group, comb_frail_cpg_assoc_val, fidmrs_cpg_combined$IlmnID)
comb_frail_cpg_assoc_prepostfuture_val = prepostfuture_data_merge(tp_future_group,
                                                                  comb_frail_cpg_assoc_val,
                                                                  fidmrs_cpg_combined$IlmnID)
```

```{r}
dmr_frail_coeff_val = lapply(1:925, function(x) {
  loci_lmer = fidmrs_cpg_combined %>% filter(dmr == unique(fidmrs_cpg_combined$dmr)[[x]])
  expr_merged = comb_frail_cpg_assoc_val %>%
    dplyr::select(loci_lmer$IlmnID, id, mouse_id, logFI, sex) %>%
    melt(., id = c("id", "mouse_id", "logFI", "sex")) %>%
    as.data.frame()
  res = lmer(logFI ~ value + as.factor(sex) + (1|mouse_id) + (1|variable), data = expr_merged)
  coef = summary(res)$coefficients
  return(c(coef[2, 1], coef[2, 5]))}) %>%
  Reduce(rbind, .) %>%
  as.data.frame()
```

```{r}
frail_coeff_dis_val = cbind(dmr_frail_coeff, dmr_frail_coeff_val) %>%
  as.data.frame() %>%
  `colnames<-`(c("coef_dis", "p_dis", "coef_val", "p_val")) %>%
  `rownames<-`(unique(fidmrs_cpg_combined$dmr))

ggplot(data = frail_coeff_dis_val) +
   geom_point(aes(x = coef_dis,
           y = coef_val),
           color = "#44AA99",
           size = 0.75) +
   geom_point(data = frail_coeff_dis_val %>% filter(coef_dis * coef_val < 0),
              aes(x = coef_dis,
                  y = coef_val),
              color = "#762A83", shape = 17) + 
   labs(x = "Coefficient (discovery cohort)",
        y = "Coefficient (validation cohort)") +
   geom_abline(slope = 1, intercept = 0, lty = 3) +
   geom_vline(xintercept = 0, lty = 3) +
   geom_hline(yintercept = 0, lty = 3) 
```


```{r, message = FALSE, warning = FALSE}
fidmr_assoc_f_val = lapply(fm_all_dmrs, function(x) {
  dmr_cpgs = fm_all_dmr_cpgs_combined %>% filter(dmr == x)
  dmr_cpgs_melt = comb_frail_cpg_assoc_val %>%
    filter(sex == "1") %>%
    dplyr::select(unique(dmr_cpgs$IlmnID), id, mouse_id, age, logFI) %>%
    melt(., id = c("id", "mouse_id", "age", "logFI")) %>%
    as.data.frame()
  dmr_assoc_res = lmer(value ~ logFI + (1|mouse_id) + (0 + logFI|mouse_id), data = dmr_cpgs_melt)
  return(dmr_assoc_res)
})

fidmr_assoc_m_val = lapply(fm_all_dmrs, function(x) {
  dmr_cpgs = fm_all_dmr_cpgs_combined %>% filter(dmr == x)
  dmr_cpgs_melt = comb_frail_cpg_assoc_val %>%
    filter(sex == "0") %>%
    dplyr::select(unique(dmr_cpgs$IlmnID), id, mouse_id, age, logFI) %>%
    melt(., id = c("id", "mouse_id", "age", "logFI")) %>%
    as.data.frame()
  dmr_assoc_res = lmer(value ~ logFI + (1|mouse_id) + (0 + logFI|mouse_id), data = dmr_cpgs_melt)
  return(dmr_assoc_res)
})
```

```{r}
fidmr_assoc_coeff_val = data.frame(dmr = fm_all_dmrs,
                                    coeff_f_dis = sapply(1:903, function(x) 
                                      {coef = summary(fidmr_assoc_f[[x]])$coefficients[2, 1] 
                                      return(coef)}),
                                    coeff_f_val = sapply(1:903, function(x) 
                                      {coef = summary(fidmr_assoc_f_val[[x]])$coefficients[2, 1] 
                                      return(coef)}),
                                    coeff_m_dis = sapply(1:903, function(x) 
                                      {coef = summary(fidmr_assoc_m[[x]])$coefficients[2, 1] 
                                      return(coef)}),
                                    coeff_m_val = sapply(1:903, function(x) 
                                      {coef = summary(fidmr_assoc_m_val[[x]])$coefficients[2, 1] 
                                      return(coef)})) 

female_spe_assoc_coeff_val = fidmr_assoc_coeff_val %>% filter(dmr %in% female_fidmrs$dmr)
male_spe_assoc_coeff_val = fidmr_assoc_coeff_val %>% filter(dmr %in% male_fidmrs$dmr)
```

```{r}
ggplot(data = female_spe_assoc_coeff_val) +
   geom_point(aes(x = coeff_f_dis,
           y = coeff_f_val),
           color = "#44AA99",
           size = 0.75) +
   geom_point(data = female_spe_assoc_coeff_val %>% filter(coeff_f_dis * coeff_f_val < 0),
              aes(x = coeff_f_dis,
                  y = coeff_f_val),
              color = "#762A83", shape = 17) + 
   labs(x = "Coefficient (discovery cohort females)",
        y = "Coefficient (validation cohort females)") +
   geom_abline(slope = 1, intercept = 0, lty = 3) +
   geom_vline(xintercept = 0, lty = 3) +
   geom_hline(yintercept = 0, lty = 3) 
```

```{r}
ggplot(data = male_spe_assoc_coeff_val) +
   geom_point(aes(x = coeff_m_dis,
           y = coeff_m_val),
           color = "#44AA99",
           size = 0.75) +
   geom_point(data = male_spe_assoc_coeff_val %>% filter(coeff_m_dis * coeff_m_val < 0),
              aes(x = coeff_m_dis,
                  y = coeff_m_val),
              color = "#762A83", shape = 17) + 
   labs(x = "Coefficient (discovery cohort males)",
        y = "Coefficient (validation cohort males)") +
   geom_abline(slope = 1, intercept = 0, lty = 3) +
   geom_vline(xintercept = 0, lty = 3) +
   geom_hline(yintercept = 0, lty = 3) 
```

```{r, warning = FALSE, message = FALSE}
dmr_assoc_res_val = dmr_assoc_frailty_outcome(comb_frail_cpg_assoc_val, comb_frail_cpg_assoc_prepost_val,
                                              comb_frail_cpg_assoc_prepostfuture_val)
dmr_assoc_fic_val_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val[[x]][[1]])$coefficients
  coef2 = summary(dmr_assoc_res_val[[x]][[2]])$coefficients
  bound1 = confint(dmr_assoc_res_val[[x]][[1]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fic",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_fif_val_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val[[x]][[3]])$coefficients
  coef2 = summary(dmr_assoc_res_val[[x]][[4]])$coefficients
  bound1 = confint(dmr_assoc_res_val[[x]][[3]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fif",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_deltafi_val_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val[[x]][[5]])$coefficients
  coef2 = summary(dmr_assoc_res_val[[x]][[6]])$coefficients
  bound1 = confint(dmr_assoc_res_val[[x]][[5]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_delta",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fic_val_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val[[x]][[7]])$coefficients
  coef2 = summary(dmr_assoc_res_val[[x]][[8]])$coefficients
  bound1 = confint(dmr_assoc_res_val[[x]][[7]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fic",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fif_val_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val[[x]][[9]])$coefficients
  coef2 = summary(dmr_assoc_res_val[[x]][[10]])$coefficients
  bound1 = confint(dmr_assoc_res_val[[x]][[9]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[7, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fif",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_deltafi_val_res = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val[[x]][[11]])$coefficients
  coef2 = summary(dmr_assoc_res_val[[x]][[12]])$coefficients
  bound1 = confint(dmr_assoc_res_val[[x]][[11]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[7, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_delta",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_val_match = list(dmr_assoc_fic_val_res$dmr[dmr_assoc_fic_val_res$dmr %in% 
                                                 dmr_assoc_fic_res$dmr],
                     dmr_assoc_fif_val_res$dmr[dmr_assoc_fif_val_res$dmr %in% 
                                                 dmr_assoc_fif_res$dmr],
                     dmr_assoc_deltafi_val_res$dmr[dmr_assoc_deltafi_val_res$dmr %in% 
                                                     dmr_assoc_deltafi_res$dmr],
                     deltadmr_assoc_fic_val_res$dmr[deltadmr_assoc_fic_val_res$dmr %in%
                                                      deltadmr_assoc_fic_res$dmr],
                     deltadmr_assoc_fif_val_res$dmr[deltadmr_assoc_fif_val_res$dmr %in%
                                                      deltadmr_assoc_fif_res$dmr],
                     deltadmr_assoc_deltafi_val_res$dmr[deltadmr_assoc_deltafi_val_res$dmr %in%
                                                          deltadmr_assoc_deltafi_res$dmr])
```

```{r}
dmr_val_match = list(dmr_assoc_fif_val_res$dmr[dmr_assoc_fif_val_res$dmr %in% 
                                                 dmr_assoc_fif_res$dmr],
                     dmr_assoc_deltafi_val_res$dmr[dmr_assoc_deltafi_val_res$dmr %in% 
                                                     dmr_assoc_deltafi_res$dmr],
                     deltadmr_assoc_fif_val_res$dmr[deltadmr_assoc_fif_val_res$dmr %in%
                                                      deltadmr_assoc_fif_res$dmr],
                     deltadmr_assoc_deltafi_val_res$dmr[deltadmr_assoc_deltafi_val_res$dmr %in%
                                                          deltadmr_assoc_deltafi_res$dmr])
```


```{r}
dmr_val_match_tbl_mix = Reduce(rbind, list(deltadmr_assoc_fif_res %>% filter(dmr %in% dmr_val_match[[3]]) %>% mutate(cohort = "dis"),
                                       deltadmr_assoc_fif_val_res %>% filter(dmr %in% dmr_val_match[[3]]),
                                       deltadmr_assoc_deltafi_res %>% filter(dmr %in% dmr_val_match[[4]]) %>% mutate(cohort = "dis"),
                                       deltadmr_assoc_deltafi_val_res %>% filter(dmr %in% dmr_val_match[[4]]))) %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type, cohort) %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ "")) 
dmr_val_match_tbl_mix_order = dmr_val_match_tbl_mix %>% filter(cohort == "dis") %>%
  arrange(type, coef)

dmr_val_match_tbl_mix$dmr = factor(dmr_val_match_tbl_mix$dmr, 
                                   levels = dmr_val_match_tbl_mix_order$dmr)
dmr_val_match_tbl_mix$cohort = factor(dmr_val_match_tbl_mix$cohort, levels = c("val", "dis"))
dmr_val_match_tbl_mix$type = factor(dmr_val_match_tbl_mix$type, levels = c("delta_delta", "delta_fif"))
pdf(file = "~/Desktop/DNA methylation/figures_0718/fig10a.pdf", height = 3.5, width = 4.5)
ggplot(data = dmr_val_match_tbl_mix) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 3.2) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type),
             shape = as.factor(cohort)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type),
                    shape = as.factor(cohort)),
                position = position_dodge(width = 0.8),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.1, 0.1) +
      scale_shape_manual(values = c(17, 16)) + 
      scale_color_manual(values = c("#DDAA33", "#762A83")) +
      theme(axis.text.y = element_text(size = 7, face = "bold"))
```

## 6.6 female validation association
```{r, warning = FALSE, message = FALSE}
comb_frail_cpg_assoc_val_female = comb_frail_cpg_assoc_val %>% filter(sex == "1")
comb_frail_cpg_assoc_prepost_val_female = comb_frail_cpg_assoc_prepost_val %>% filter(sex.x == "1")
comb_frail_cpg_assoc_prepostfuture_val_female = comb_frail_cpg_assoc_prepostfuture_val %>% filter(sex == "1")
dmr_assoc_res_val_female = dmr_assoc_frailty_outcome_sex(comb_frail_cpg_assoc_val_female,
                                                         comb_frail_cpg_assoc_prepost_val_female,
                                                         comb_frail_cpg_assoc_prepostfuture_val_female)

dmr_assoc_fic_val_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_female[[x]][[1]])$coefficients
  coef2 = summary(dmr_assoc_res_val_female[[x]][[2]])$coefficients
  bound1 = confint(dmr_assoc_res_val_female[[x]][[1]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[4, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fic",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_fif_val_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_female[[x]][[3]])$coefficients
  coef2 = summary(dmr_assoc_res_val_female[[x]][[4]])$coefficients
  bound1 = confint(dmr_assoc_res_val_female[[x]][[3]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fif",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_deltafi_val_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_female[[x]][[5]])$coefficients
  coef2 = summary(dmr_assoc_res_val_female[[x]][[6]])$coefficients
  bound1 = confint(dmr_assoc_res_val_female[[x]][[5]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_delta",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fic_val_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_female[[x]][[7]])$coefficients
  coef2 = summary(dmr_assoc_res_val_female[[x]][[8]])$coefficients
  bound1 = confint(dmr_assoc_res_val_female[[x]][[7]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fic",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fif_val_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_female[[x]][[9]])$coefficients
  coef2 = summary(dmr_assoc_res_val_female[[x]][[10]])$coefficients
  bound1 = confint(dmr_assoc_res_val_female[[x]][[9]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fif",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_deltafi_val_res_female = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_female[[x]][[11]])$coefficients
  coef2 = summary(dmr_assoc_res_val_female[[x]][[12]])$coefficients
  bound1 = confint(dmr_assoc_res_val_female[[x]][[11]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_delta",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_val_female_match = list(dmr_assoc_fic_val_res_female$dmr[dmr_assoc_fic_val_res_female$dmr %in% 
                                                               dmr_assoc_fic_res_female$dmr],
                            dmr_assoc_fif_val_res_female$dmr[dmr_assoc_fif_val_res_female$dmr %in% 
                                                               dmr_assoc_fif_res_female$dmr],
                            dmr_assoc_deltafi_val_res_female$dmr[dmr_assoc_deltafi_val_res_female$dmr %in%
                                                                   dmr_assoc_deltafi_res_female$dmr],
                            deltadmr_assoc_fic_val_res_female$dmr[deltadmr_assoc_fic_val_res_female$dmr %in%
                                                                    deltadmr_assoc_fic_res_female$dmr],
                            deltadmr_assoc_fif_val_res_female$dmr[deltadmr_assoc_fif_val_res_female$dmr %in%
                                                                    deltadmr_assoc_fif_res_female$dmr],
                            deltadmr_assoc_deltafi_val_res_female$dmr[deltadmr_assoc_deltafi_val_res_female$dmr %in%
                                                                        deltadmr_assoc_deltafi_res_female$dmr])
```

```{r}
dmr_val_female_match = list(dmr_assoc_fif_val_res_female$dmr[dmr_assoc_fif_val_res_female$dmr %in% 
                                                               dmr_assoc_fif_res_female$dmr],
                            dmr_assoc_deltafi_val_res_female$dmr[dmr_assoc_deltafi_val_res_female$dmr %in%
                                                                   dmr_assoc_deltafi_res_female$dmr],
                            deltadmr_assoc_fif_val_res_female$dmr[deltadmr_assoc_fif_val_res_female$dmr %in%
                                                                    deltadmr_assoc_fif_res_female$dmr],
                            deltadmr_assoc_deltafi_val_res_female$dmr[deltadmr_assoc_deltafi_val_res_female$dmr %in%
                                                                        deltadmr_assoc_deltafi_res_female$dmr])
```


```{r}
dmr_val_match_tbl_female = Reduce(rbind, 
                                  list(dmr_assoc_deltafi_res_female %>% filter(dmr %in% dmr_val_female_match[[2]]) %>% mutate(cohort = "dis"),
                                       dmr_assoc_deltafi_val_res_female %>% filter(dmr %in% dmr_val_female_match[[2]]),
                                       deltadmr_assoc_deltafi_res_female %>% filter(dmr %in% dmr_val_female_match[[4]]) %>% mutate(cohort = "dis"),
                                       deltadmr_assoc_deltafi_val_res_female %>% filter(dmr %in% dmr_val_female_match[[4]]))) %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type, cohort) %>%
  as.data.frame() %>%
  mutate(sig = case_when(
                         adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ "")) 
dmr_val_match_arrange_female = dmr_val_match_tbl_female %>% filter(cohort == "dis") %>% arrange(type, coef)
dmr_val_match_tbl_female$dmr = factor(dmr_val_match_tbl_female$dmr, levels = unique(dmr_val_match_arrange_female$dmr))
dmr_val_match_tbl_female$cohort = factor(dmr_val_match_tbl_female$cohort, levels = c("val", "dis"))
dmr_val_match_tbl_female$type = factor(dmr_val_match_tbl_female$type, levels = c("delta_delta", "dmr_delta"))

ggplot(data = dmr_val_match_tbl_female) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 4.8) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type),
             shape = as.factor(cohort)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type),
                    shape = as.factor(cohort)),
                position = position_dodge(width = 0.8),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.08, 0.15) +
      scale_shape_manual(values = c(17, 16)) + 
      scale_color_manual(values = c("#DDAA33", "#117777")) +
      theme(axis.text.y = element_text(size = 8, face = "bold"))
```

## 6.7 male validation association
```{r, message = FALSE, warning = FALSE}
comb_frail_cpg_assoc_val_male = comb_frail_cpg_assoc_val %>% filter(sex == "0")
comb_frail_cpg_assoc_prepost_val_male = comb_frail_cpg_assoc_prepost_val %>% filter(sex.x == "0")
comb_frail_cpg_assoc_prepostfuture_val_male = comb_frail_cpg_assoc_prepostfuture_val %>% filter(sex == "0")
dmr_assoc_res_val_male = dmr_assoc_frailty_outcome_sex(comb_frail_cpg_assoc_val_male,
                                                         comb_frail_cpg_assoc_prepost_val_male,
                                                         comb_frail_cpg_assoc_prepostfuture_val_male)
load("~/Desktop/DNA methylation/dmr_assoc_res_val_male.RData")
dmr_assoc_fic_val_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_male[[x]][[1]])$coefficients
  coef2 = summary(dmr_assoc_res_val_male[[x]][[2]])$coefficients
  bound1 = confint(dmr_assoc_res_val_male[[x]][[1]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[4, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fic",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_fif_val_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_male[[x]][[3]])$coefficients
  coef2 = summary(dmr_assoc_res_val_male[[x]][[4]])$coefficients
  bound1 = confint(dmr_assoc_res_val_male[[x]][[3]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_fif",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_assoc_deltafi_val_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_male[[x]][[5]])$coefficients
  coef2 = summary(dmr_assoc_res_val_male[[x]][[6]])$coefficients
  bound1 = confint(dmr_assoc_res_val_male[[x]][[5]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "dmr_delta",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fic_val_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_male[[x]][[7]])$coefficients
  coef2 = summary(dmr_assoc_res_val_male[[x]][[8]])$coefficients
  bound1 = confint(dmr_assoc_res_val_male[[x]][[7]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[5, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fic",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_fif_val_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_male[[x]][[9]])$coefficients
  coef2 = summary(dmr_assoc_res_val_male[[x]][[10]])$coefficients
  bound1 = confint(dmr_assoc_res_val_male[[x]][[9]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_fif",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

deltadmr_assoc_deltafi_val_res_male = lapply(1:925, function(x) {
  coef1 = summary(dmr_assoc_res_val_male[[x]][[11]])$coefficients
  coef2 = summary(dmr_assoc_res_val_male[[x]][[12]])$coefficients
  bound1 = confint(dmr_assoc_res_val_male[[x]][[11]], method = "Wald")
  return(c(coef1[2, 1], coef1[2, 5], bound1[5, 1], bound1[5, 2], coef2[6, 5]))
}) %>% Reduce(rbind, .) %>% as.data.frame() %>%
  `colnames<-`(c("coef", "p1", "lb", "ub", "p2")) %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr),
         adjp1 = p.adjust(p1, method = "fdr"),
         adjp2 = p.adjust(p2, method = "fdr"),
         type = "delta_delta",
         cohort = "val") %>%
  filter(adjp1 < 0.05 & adjp2 >= 0.05)

dmr_val_male_match = list(dmr_assoc_fic_val_res_male$dmr[dmr_assoc_fic_val_res_male$dmr %in% dmr_assoc_fic_res_male$dmr],
                            dmr_assoc_fif_val_res_male$dmr[dmr_assoc_fif_val_res_male$dmr %in% dmr_assoc_fif_res_male$dmr],
                            dmr_assoc_deltafi_val_res_male$dmr[dmr_assoc_deltafi_val_res_male$dmr %in% dmr_assoc_deltafi_res_male$dmr],
                            deltadmr_assoc_fic_val_res_male$dmr[deltadmr_assoc_fic_val_res_male$dmr %in% deltadmr_assoc_fic_res_male$dmr],
                            deltadmr_assoc_fif_val_res_male$dmr[deltadmr_assoc_fif_val_res_male$dmr %in% deltadmr_assoc_fif_res_male$dmr],
                            deltadmr_assoc_deltafi_val_res_male$dmr[deltadmr_assoc_deltafi_val_res_male$dmr %in% deltadmr_assoc_deltafi_res_male$dmr])
```

```{r}
dmr_val_male_match = list(
                            dmr_assoc_fif_val_res_male$dmr[dmr_assoc_fif_val_res_male$dmr %in% dmr_assoc_fif_res_male$dmr],
                            dmr_assoc_deltafi_val_res_male$dmr[dmr_assoc_deltafi_val_res_male$dmr %in% dmr_assoc_deltafi_res_male$dmr],
                          
                            deltadmr_assoc_fif_val_res_male$dmr[deltadmr_assoc_fif_val_res_male$dmr %in% deltadmr_assoc_fif_res_male$dmr],
                            deltadmr_assoc_deltafi_val_res_male$dmr[deltadmr_assoc_deltafi_val_res_male$dmr %in% deltadmr_assoc_deltafi_res_male$dmr])
```

```{r}
dmr_val_match_tbl_male = Reduce(rbind, 
                                list(
                                     deltadmr_assoc_fif_res_male %>% filter(dmr %in% dmr_val_male_match[[3]]) %>% mutate(cohort = "dis"),
                                     deltadmr_assoc_fif_val_res_male %>% filter(dmr %in% dmr_val_male_match[[3]]),
                                     deltadmr_assoc_deltafi_res_male %>% filter(dmr %in% dmr_val_male_match[[4]]) %>% mutate(cohort = "dis"),
                                     deltadmr_assoc_deltafi_val_res_male %>% filter(dmr %in% dmr_val_male_match[[4]]))) %>%
  dplyr::select(coef, adjp1, lb, ub, dmr, type, cohort) %>%
  as.data.frame() %>%
  mutate(sig = case_when(adjp1 < 0.001 ~ "***",
                         adjp1 < 0.01 ~ "**",
                         adjp1 < 0.05 ~ "*",
                         adjp1 < 0.1 ~ "",
                         TRUE ~ "")) 

dmr_val_match_arrange_male = dmr_val_match_tbl_male %>% 
  filter(cohort == "dis") %>% 
  arrange(factor(type, levels = c("delta_delta", "delta_fif")), coef)
dmr_val_match_tbl_male$dmr = factor(dmr_val_match_tbl_male$dmr, levels = unique(dmr_val_match_arrange_male$dmr))
dmr_val_match_tbl_male$cohort = factor(dmr_val_match_tbl_male$cohort, levels = c("val", "dis"))
dmr_val_match_tbl_male$type = factor(dmr_val_match_tbl_male$type, levels = c("delta_delta", "delta_fif"))
ggplot(data = dmr_val_match_tbl_male) +
  geom_hline(aes(yintercept = dmr), color = "grey95", size = 4.5) +
       ggstance::geom_pointrangeh(
         aes(x = as.numeric(coef),
             y = dmr,
             xmin = as.numeric(lb),
             xmax = as.numeric(ub),
             color = as.factor(type),
             shape = as.factor(cohort)),
             size = 0.25,
         position = position_dodge(width = 0.8))  +
      geom_vline(xintercept = 0, lty = 3) +
      labs(x = "Coefficient (95% CI)",
           y = "",
           color = "Frailty outcomes") +
      geom_text(aes(y = dmr,
                    x = Inf,
                    label = sig,
                    color = as.factor(type),
                    shape = as.factor(cohort)),
                position = position_dodge(width = 0.8),
                hjust = 1, vjust = 0.9, size = 5,
                show.legend = FALSE) +
      xlim(-0.13, 0.18) +
      scale_shape_manual(values = c(17, 16)) + 
      scale_color_manual(values = c("#DDAA33", "#762A83")) +
      theme(axis.text.y = element_text(size = 8, face = "bold"))
```

```{r}
validation_dmr_all = c(Reduce(c, dmr_val_match) %>% unique(), 
                   Reduce(c, dmr_val_female_match) %>% unique(),
                   Reduce(c, dmr_val_male_match) %>% unique()) %>%
  unique()

validation_dmrs_all_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% validation_dmr_all)
validation_dmrs_cpgs_genes = prob_enriched_genes(validation_dmrs_all_cpgs$IlmnID)
validation_dmrs_cpgs_paths = prob_enriched_path(validation_dmrs_all_cpgs$IlmnID)
```

# 7. cpgs within dmrs variability/entropy
## 7.1 variability in sex-inclusive and -specific subgroups
```{r, warning = FALSE,  message = FALSE}
## inter-individual variance of slope and intercept 
## methy_dat_dis_meta_update
methy_dat_dis_M_excl = methy_dat_dis_meta_update %>%
  filter(Time.label != "T5") %>%
  dplyr::select(mix_fidmr_cpgs$IlmnID, logFI, Sex, mouse_id)

frail_cpg_assoc = lapply(1:3337, function(x) {
  models = lmer(methy_dat_dis_M_excl[, x] ~ methy_dat_dis_M_excl$logFI + methy_dat_dis_M_excl$Sex + (1| methy_dat_dis_M_excl$mouse_id) + (0 + methy_dat_dis_M_excl$logFI|methy_dat_dis_M_excl$mouse_id))
})

mix_fidmr_cpgs_cate = merge(mix_fidmr_cpgs, mix_fidmr, by = "dmr", all.x = TRUE) %>%
  as.data.frame()

frail_cpg_coeff_random = data.frame(IlmnID = mix_fidmr_cpgs$IlmnID,
                                    cate = mix_fidmr_cpgs_cate$cate,
                                    vc_inter = sapply(1:3337, function(y) 
                                      {vc1 = as.data.frame(VarCorr(frail_cpg_assoc[[y]]))[1, 4] 
                                      return(vc1)}),
                                    vc_slope = sapply(1:3337, function(z) 
                                      {vc2 = as.data.frame(VarCorr(frail_cpg_assoc[[z]]))[2, 4] 
                                      return(vc2)})) 
frail_cpg_coeff_random$cate = factor(frail_cpg_coeff_random$cate, levels = c("frail", "dual", "age"))
vc_inter_plot = ggplot(data = frail_cpg_coeff_random,
       aes(x = as.factor(cate),
           y = vc_inter)) +
   geom_violin(aes(color = cate)) +
   geom_boxplot(aes(color = cate), width = 0.1, outlier.shape = NA) +
   scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
   labs(x = "")

vc_slope_plot = ggplot(data = frail_cpg_coeff_random,
       aes(x = as.factor(cate),
           y = vc_slope)) +
   geom_violin(aes(color = cate)) +
   geom_boxplot(aes(color = cate), width = 0.1, outlier.shape = NA) +
   scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
   labs(x = "")
```

```{r}
methy_dat_dis_M_f = methy_dat_dis_meta_update %>%
  filter(Sex == "female") %>%
  dplyr::select(female_fidmrs_cpgs$IlmnID, logFI, mouse_id)

frail_cpg_assoc_f = lapply(1:6091, function(x) {
  models = lmer(methy_dat_dis_M_f[, x] ~ methy_dat_dis_M_f$logFI + (1| methy_dat_dis_M_f$mouse_id) + (0 + methy_dat_dis_M_f$logFI|methy_dat_dis_M_f$mouse_id))
})

female_fidmr_cpgs_cate = merge(female_fidmrs_cpgs, female_fidmrs, by = "dmr", all.x = TRUE) %>%
  as.data.frame()

frail_cpg_coeff_random_f = data.frame(IlmnID = female_fidmrs_cpgs$IlmnID,
                                    cate = female_fidmr_cpgs_cate$dmr_cate,
                                    vc_inter = sapply(1:6091, function(y) 
                                      {vc1 = as.data.frame(VarCorr(frail_cpg_assoc_f[[y]]))[1, 4] 
                                      return(vc1)}),
                                    vc_slope = sapply(1:6091, function(z) 
                                      {vc2 = as.data.frame(VarCorr(frail_cpg_assoc_f[[z]]))[2, 4] 
                                      return(vc2)})) 

frail_cpg_coeff_random_f$cate = factor(frail_cpg_coeff_random_f$cate, levels = c("dual", "age"))
vc_inter_f_plot = ggplot(data = frail_cpg_coeff_random_f,
       aes(x = as.factor(cate),
           y = vc_inter)) +
   geom_violin(aes(color = cate)) +
   geom_boxplot(aes(color = cate), width = 0.1, outlier.shape = NA) +
   scale_color_manual(values = c("#DDAA33", "#44AA99")) +
   labs(x = "")

vc_slope_f_plot = ggplot(data = frail_cpg_coeff_random_f,
       aes(x = as.factor(cate),
           y = vc_slope)) +
   geom_violin(aes(color = cate)) +
   geom_boxplot(aes(color = cate), width = 0.1, outlier.shape = NA) +
   scale_color_manual(values = c("#DDAA33", "#44AA99")) +
   labs(x = "")
```

```{r}
methy_dat_dis_M_m = methy_dat_dis_meta_update %>%
  filter(Sex == "male") %>%
  dplyr::select(male_fidmrs_cpgs$IlmnID, logFI, mouse_id)

frail_cpg_assoc_m = lapply(1:3000, function(x) {
  models = lmer(methy_dat_dis_M_m[, x] ~ methy_dat_dis_M_m$logFI + (1| methy_dat_dis_M_m$mouse_id) + (0 + methy_dat_dis_M_m$logFI|methy_dat_dis_M_m$mouse_id))
})

male_fidmr_cpgs_cate = merge(male_fidmrs_cpgs, male_fidmrs, by = "dmr", all.x = TRUE) %>%
  as.data.frame()

frail_cpg_coeff_random_m = data.frame(IlmnID = male_fidmrs_cpgs$IlmnID,
                                    cate = male_fidmr_cpgs_cate$cate,
                                    vc_inter = sapply(1:3000, function(y) 
                                      {vc1 = as.data.frame(VarCorr(frail_cpg_assoc_m[[y]]))[1, 4] 
                                      return(vc1)}),
                                    vc_slope = sapply(1:3000, function(z) 
                                      {vc2 = as.data.frame(VarCorr(frail_cpg_assoc_m[[z]]))[2, 4] 
                                      return(vc2)})) 

frail_cpg_coeff_random_m$cate = factor(frail_cpg_coeff_random_m$cate, levels = c("frail", "dual", "age"))
vc_inter_m_plot = ggplot(data = frail_cpg_coeff_random_m,
       aes(x = as.factor(cate),
           y = vc_inter)) +
   geom_violin(aes(color = cate)) +
   geom_boxplot(aes(color = cate), width = 0.1, outlier.shape = NA) +
   scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
   labs(x = "")

vc_slope_m_plot = ggplot(data = frail_cpg_coeff_random_m,
       aes(x = as.factor(cate),
           y = vc_slope)) +
   geom_violin(aes(color = cate)) +
   geom_boxplot(aes(color = cate), width = 0.1, outlier.shape = NA) +
   scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
   labs(x = "")
```

## 7.2 interslop variation examples
```{r}
interslop_example_tbl = cbind(methy_dat_dis_meta_excl %>% 
                                     dplyr::select(id, mouse_id, FI.Score, Sex),
                            methy_dat_dis_M_excl %>% 
                                     dplyr::select(cg37536990_TC11, cg42178351_BC11)) %>%
  as.data.frame() %>%
  filter(!mouse_id %in% c("52.2", "68.3", "71.4")) %>%
  melt(., id = c("id", "mouse_id", "FI.Score", "Sex")) %>%
  as.data.frame()

ggplot(interslop_example_tbl, 
       aes(x = log2(FI.Score), 
           y = value)) +
  geom_point(aes(color = Sex), size = 0.5) +
  geom_smooth(method = "lm", aes(group = as.factor(mouse_id), color = Sex), se = FALSE, alpha = 0.4) +
  scale_color_manual(values = c("#edc5d2", "#b8cde3")) +
  new_scale_color() +
  geom_smooth(method = "lm", aes(group = Sex, color = Sex, fill = Sex), se = TRUE, alpha = 0.2) +
  scale_color_manual(values = c("#AA4466", "#4477AA")) +
  facet_wrap(.~ variable)
```

## 7.3 test of vmps
```{r}
## VMP analysis of frail-related CpGs
methy_fm_excl_resid = residuals(fm_frail_fit2_res, methy_counts_excl)
fm_excl_frail_cpg_resid = methy_fm_excl_resid[rownames(methy_fm_excl_resid) %in% mix_fidmr_cpgs$IlmnID, ]
methy_normal_test = data.frame(loci = rownames(fm_excl_frail_cpg_resid),
                               shapiro_p = sapply(1:nrow(fm_excl_frail_cpg_resid), function(x) {shapiro.test(fm_excl_frail_cpg_resid[x, ])$p.value})) %>%
  filter(shapiro_p >= 0.05)
methy_fm_excl_vmp_assoc = data.frame(loci = methy_normal_test$loci, 
                                     bp_p = sapply(1:nrow(methy_normal_test), function(x) {
                                     cpg = methy_normal_test[x, 1]
                                     lm_res = lm((methy_fm_excl_resid[cpg, ])^2 ~ frail)
                                     return(bptest(lm_res)$p.value)})) %>%
  mutate(adjp = p.adjust(bp_p, method = "BH")) %>%
  filter(adjp < 0.1)
```

```{r}
female_methy_frail_cpg = female_residuals[rownames(female_residuals) %in% female_fidmrs_cpgs$IlmnID, ]
female_methy_normal_test = data.frame(loci = rownames(female_methy_frail_cpg),
                               shapiro_p = sapply(1:nrow(female_methy_frail_cpg), function(x) {shapiro.test(female_methy_frail_cpg[x, ])$p.value})) %>%
  filter(shapiro_p >= 0.05)
female_methy_vmp_assoc = data.frame(loci = female_methy_normal_test$loci, 
                                    bp_p = sapply(1:nrow(female_methy_normal_test), function(x) {
                                    cpg = female_methy_normal_test[x, 1]
                                    lm_res = lm((female_methy_frail_cpg[cpg, ])^2 ~ frail_female)
                                       return(bptest(lm_res)$p.value)})) %>%
  mutate(adjp = p.adjust(bp_p, method = "BH")) %>% filter(adjp < 0.1)
```

```{r}
male_methy_frail_cpg = male_residuals[rownames(male_residuals) %in% male_fidmrs_cpgs$IlmnID, ]
male_methy_normal_test = data.frame(loci = rownames(male_methy_frail_cpg),
                               shapiro_p = sapply(1:nrow(male_methy_frail_cpg), function(x) {shapiro.test(male_methy_frail_cpg[x, ])$p.value})) %>%
  filter(shapiro_p >= 0.05)
male_methy_vmp_assoc = data.frame(loci = male_methy_normal_test$loci, 
                                  bp_p = sapply(1:nrow(male_methy_normal_test), function(x) {
                                  cpg = male_methy_normal_test[x, 1]
                                  lm_res = lm((male_methy_frail_cpg[cpg, ])^2 ~ frail_male)
                                       return(bptest(lm_res)$p.value)})) %>%
  mutate(adjp = p.adjust(bp_p, method = "BH")) %>% filter(adjp < 0.1)
```

## 7.4 male VMPs classification
```{r}
## SNR analysis to determine stochasticity
methy_dat_dis_meta_m_traject = methy_dat_dis_meta %>%
  filter(Sex == "male") %>%
  mutate(mouse_id = Mouse.ID) %>%
  filter(!mouse_id %in% c("52.2", "68.3", "71.4")) %>%
  mutate(logFI = log2(FI.Score))
  
methy_dat_dis_M_m_traject = methy_dat_dis_M %>%
  filter(rownames(.) %in% methy_dat_dis_meta_m_traject$id)
#vmp classification
#SNR trajectory
#SNR cloud
# exclude 52.2, 68.3, and 71.4. Only contains 1 time point
snr_ratio = sapply(male_methy_vmp_assoc$loci, function(x) {
  M_value = methy_dat_dis_M_m_traject[, x]
  adj_beta = (2^M_value)/(1 + 2^M_value)
  model = lm(adj_beta ~ methy_dat_dis_meta_m_traject$logFI)
  each_indv = methy_dat_dis_meta_m_traject %>%
    as.data.frame() %>%
    mutate(cpg = adj_beta,
           deviation = abs(unname(residuals(model))),
           fitted = unname(predict(model)))
  snr_z = sapply(unique(each_indv$mouse_id), function(y) {
    each_indv_z = each_indv %>% filter(mouse_id == y) %>%
      mutate(snr = fitted/deviation)
    return(mean(each_indv_z$snr))
  })
  each_indv_tp = each_indv %>%
    dplyr::group_by(Time.label) %>%
    dplyr::summarise(mean_M = mean(cpg))
  each_indv_update = merge(each_indv, each_indv_tp, by = "Time.label") %>%
    as.data.frame() %>%
    mutate(snr_cloud = mean_M/abs(cpg - mean_M))
  snr_cloud = mean(each_indv_update$snr_cloud)
  ratio = mean(snr_z)/snr_cloud
  return(ratio)
})
```

```{r}
determ_cpg_snr_res = names(snr_ratio[snr_ratio >10])
stoch_cpg_snr_res = names(snr_ratio[snr_ratio < 1])
```

## 7.5 male VMPs determinstric examples
```{r}
## vmp examples using smoothed lines
male_vmp_example_tbl1 = cbind(methy_dat_dis_meta_excl %>% 
                                     dplyr::select(id, mouse_id, FI.Score, Sex),
                            methy_dat_dis_M_excl %>% 
                                     dplyr::select(cg43392154_TC21)) %>%
  as.data.frame() %>%
  filter(!mouse_id %in% c("52.2", "68.3", "71.4")) %>%
  filter(Sex == "male")

male_vmp_example_tbl2 = cbind(methy_dat_dis_meta_excl %>% 
                                     dplyr::select(id, mouse_id, FI.Score, Sex),
                            methy_dat_dis_M_excl %>% 
                                     dplyr::select(cg47463455_BC11)) %>%
  as.data.frame() %>% 
  filter(!mouse_id %in% c("52.2", "68.3", "71.4")) %>%
  filter(Sex == "male")

ggplot(male_vmp_example_tbl1, 
       aes(x = log2(FI.Score), 
           y = cg43392154_TC21)) +
  geom_smooth(method = "loess", aes(group = mouse_id), 
              span = 0.6, se = FALSE,
              color = "#b8cde3") +
  geom_point(size = 0.5, color = "#4477AA") +
  geom_smooth(method = "loess", aes(group = 1), alpha = 0.2, span = 0.4) +
  scale_color_manual(values = "#4477AA") 
```

```{r}
ggplot(male_vmp_example_tbl2, 
       aes(x = log2(FI.Score), 
           y = cg47463455_BC11)) +
  geom_smooth(method = "loess", aes(group = mouse_id), 
              span = 0.6, se = FALSE,
              color = "#b8cde3") +
  geom_point(size = 0.5, color = "#4477AA") +
  geom_smooth(method = "loess", aes(group = 1), alpha = 0.2, span = 0.4) +
  scale_color_manual(values = "#4477AA") 
```

## 7.6. Shannon entropy of categories of dmrs
```{r}
methy_dat_dis_meta_excl = methy_dat_dis_meta %>% 
  as.data.frame() %>% 
  mutate(logFI = log2(FI.Score),
         mouse_id = as.factor(Mouse.ID)) %>%
  filter(Time.label != "T5")
methy_dat_dis_meta_excl = methy_dat_dis_meta_excl %>%
  mutate(Sex = as.factor(Sex))
methy_dat_dis_M_excl = methy_dat_dis_M %>% filter(rownames(.) %in% methy_dat_dis_meta_excl$id)

fm_ficpg_entropy = lapply(c("frail", "dual", "age"), function(x) {
  cpg_lst = mix_fidmr_cpgs_cate[mix_fidmr_cpgs_cate$cate == x, ]
  entropy = lapply(cpg_lst$IlmnID, function(y) {
    resid = residuals(lm(methy_dat_dis_M_excl[, y] ~ as.factor(methy_dat_dis_meta_excl$Sex)))
    mean = mean(methy_dat_dis_M_excl[, y])
    adj_M = mean + resid
    adj_beta = (2^adj_M)/(1 + 2^adj_M)
    adj_beta_cal = adj_beta*log2(adj_beta) + (1 - adj_beta)*log2(1 - adj_beta)
    return(adj_beta_cal)
    }) %>% 
    Reduce(cbind, .) %>%
    as.data.frame() %>%
    rowSums()
  return(entropy)
})

methy_dat_dis_meta_entropy = methy_dat_dis_meta_excl %>%
  mutate(frail_entropy = unname(-fm_ficpg_entropy[[1]]/335),
         dual_entropy = unname(-fm_ficpg_entropy[[2]]/1514),
         age_entropy = unname(-fm_ficpg_entropy[[3]]/1488))

methy_dat_dis_meta_entropy_visual = methy_dat_dis_meta_entropy %>%
  dplyr::select(id, logFI, frail_entropy, dual_entropy, age_entropy) %>%
  melt(., id = c("logFI", "id")) %>%
  as.data.frame()
methy_dat_dis_meta_entropy_visual$variable = factor(methy_dat_dis_meta_entropy_visual$variable,
                                                 levels = c("frail_entropy", "dual_entropy", "age_entropy"))

ggplot(data = methy_dat_dis_meta_entropy_visual,
       aes(x = logFI,
           y = value,
           color = variable)) + 
  geom_point(size = 0.5) +
  geom_smooth(aes(fill = variable), formula = y ~ x, method = "lm", linewidth = 1.25) +
  scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
  scale_fill_manual(values = c("#762A83", "#DDAA33", "#44AA99"))
```

```{r}
methy_dat_dis_meta_f = methy_dat_dis_meta %>% 
  as.data.frame() %>% 
  mutate(logFI = log2(FI.Score),
         mouse_id = as.factor(Mouse.ID)) %>%
  filter(Sex == "female")
methy_dat_dis_f = methy_dat_dis %>% filter(rownames(.) %in% methy_dat_dis_meta_f$id)

f_ficpg_entropy =  lapply(c("dual", "age"), function(x) {
  cpg_lst = female_fidmr_cpgs_cate[female_fidmr_cpgs_cate$dmr_cate == x, ]
  entropy = lapply(cpg_lst$IlmnID, function(y) {
    betas = methy_dat_dis_f[, y]
    beta_cal = betas*log2(betas) + (1 - betas)*log2(1 - betas)
    return(beta_cal)
    }) %>% 
    Reduce(cbind, .) %>%
    as.data.frame() %>%
    rowSums()
  return(entropy)
})

methy_dat_dis_meta_m = methy_dat_dis_meta %>% 
  as.data.frame() %>%
  filter(Sex == "male") %>%
  mutate(logFI = log2(FI.Score),
         mouse_id = as.factor(Mouse.ID))
methy_dat_dis_m = methy_dat_dis %>% filter(rownames(.) %in% methy_dat_dis_meta_m$id)

m_ficpg_entropy = lapply(c("frail", "dual", "age"), function(x) {
  cpg_lst = male_fidmr_cpgs_cate[male_fidmr_cpgs_cate$cate == x, ]
  entropy = lapply(cpg_lst$IlmnID, function(y) {
    betas = methy_dat_dis_m[, y]
    beta_cal = betas*log2(betas) + (1 - betas)*log2(1 - betas)
    return(beta_cal)
    }) %>% 
    Reduce(cbind, .) %>%
    as.data.frame() %>%
    rowSums()
  return(entropy)
})

methy_dat_dis_meta_entropy_f = methy_dat_dis_meta_f %>%
  mutate(dual_entropy = unname(-f_ficpg_entropy[[1]]/468),
         age_entropy = unname(-f_ficpg_entropy[[2]]/5623))

methy_dat_dis_meta_entropy_m = methy_dat_dis_meta_m %>%
  mutate(frail_entropy = unname(-m_ficpg_entropy[[1]]/178),
         dual_entropy = unname(-m_ficpg_entropy[[2]]/476),
         age_entropy = unname(-m_ficpg_entropy[[3]]/2346))
```

```{r}
entropy_female_melt = methy_dat_dis_meta_entropy_f %>%
  dplyr::select(id, logFI, dual_entropy, age_entropy) %>%
  melt(., id = c("id", "logFI")) %>%
  as.data.frame()
entropy_female_melt$variable = factor(entropy_female_melt$variable, levels = c("dual_entropy", "age_entropy"))
ggplot(data = entropy_female_melt,
       aes(x = logFI,
           y = value,
           color = variable)) +
  geom_point(size = 0.5) +
  geom_smooth(aes(fill = variable), formula = y ~ x, method = "lm", linewidth = 1.25) +
  scale_color_manual(values = c("#DDAA33", "#44AA99")) +
  scale_fill_manual(values = c("#DDAA33", "#44AA99"))
```

```{r}
entropy_male_melt = methy_dat_dis_meta_entropy_m %>%
  dplyr::select(id, logFI, frail_entropy, dual_entropy, age_entropy) %>%
  melt(., id = c("id", "logFI")) %>%
  as.data.frame()
entropy_male_melt$variable = factor(entropy_male_melt$variable, levels = c("frail_entropy", "dual_entropy", "age_entropy"))
ggplot(data = entropy_male_melt,
       aes(x = logFI,
           y = value,
           color = variable)) +
  geom_point(size = 0.5) +
  geom_smooth(aes(fill = variable), formula = y ~ x, method = "lm", linewidth = 1.25) +
  scale_color_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
  scale_fill_manual(values = c("#762A83", "#DDAA33", "#44AA99")) +
  labs(y = "")
```

# 8. epigenetic frailty clock
## 8.1 select top n features 
```{r}
#load("methy_dat_tuning.RData")
dmr_coeff_rank = dmr_frail_coeff %>%
  mutate(dmr = unique(fidmrs_cpg_combined$dmr)) %>%
  arrange(desc(abs(V1)))

methy_dat_dis_merge = methy_dat_dis_M %>% 
  dplyr::select(unlist(fidmrs_cpg_combined$IlmnID)) %>%
  mutate(id = rownames(.)) %>%
  merge(., methy_dat_dis_meta, by = "id") %>%
  mutate(sex_bin = ifelse(Sex == "female", 1, 0),
         logFI = log2(FI.Score),
         age = `Age.at.assesment..days.`)

methy_dat_val_merge = methy_dat_val_M %>% 
  dplyr::select(unlist(fidmrs_cpg_combined$IlmnID)) %>%
  mutate(id = rownames(.)) %>%
  merge(., methy_dat_val_meta, by = "id") %>%
  mutate(sex_bin = ifelse(Sex == "female", 1, 0),
         logFI = log2(FI.Score),
         age = `Age.at.assesment..days.`)
methy_dat_FI_model = lm(logFI ~ age + as.factor(sex_bin), data = methy_dat_dis_merge)

select_top_n = function(n) {
  top_n_dmr = dmr_coeff_rank$dmr[1:n]
  top_n_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% top_n_dmr)
  methy_dat_dis_cpg_associated = methy_dat_dis_merge %>%
    dplyr::select(logFI, sex_bin, age, top_n_dmr_cpgs$IlmnID)
  methy_dat_dis_cpg_associated_woage = methy_dat_dis_merge %>%
    dplyr::select(logFI, sex_bin, top_n_dmr_cpgs$IlmnID)
  train_y_1 = methy_dat_dis_cpg_associated[, 1]
  train_x_1 = methy_dat_dis_cpg_associated[, -1]
  train_y_2 = methy_dat_dis_cpg_associated_woage[, 1]
  train_x_2 = methy_dat_dis_cpg_associated_woage[, -1]
  model1 = glmnet_cv(train_y_1, train_x_1, methy_dat_dis_cpg_associated)
  model2 = glmnet_cv(train_y_2, train_x_2, methy_dat_dis_cpg_associated_woage)
  frail_pred_train = methy_dat_dis_merge %>% 
                    as.data.frame() %>%
                    mutate(lm_pred = unname(predict(methy_dat_FI_model, methy_dat_dis_merge[, c("age", "sex_bin")])),
                           cpg_pred = predict(model1, as.matrix(methy_dat_dis_cpg_associated[, -1]))[, 1],
                           cpg_age_pred = predict(model2, as.matrix(methy_dat_dis_cpg_associated_woage[, -1]))[, 1]) 
  rmse_lst = c(RMSE(frail_pred_train$lm_pred, frail_pred_train$logFI),
               RMSE(frail_pred_train$cpg_pred, frail_pred_train$logFI),
               RMSE(frail_pred_train$cpg_age_pred, frail_pred_train$logFI))
  return(rmse_lst)  
}

glmnet_cv = function(train_y, train_x, df) {
  set.seed(2024)
  seeds_cv = vector(mode = "list", length = 100)
  for (i in 1:100) seeds_cv[[i]] = sample.int(10000, 400)
  seeds_cv[[101]] = sample.int(10000, 1)
  repeatedcv = trainControl(method = "repeatedcv", 
                            number = 5,
                            repeats = 20, 
                            seeds = seeds_cv)
  elastic_tuning_cpg = caret::train(x = train_x,
                              y = train_y,
                              data = df,
                              method = "glmnet",
                              trControl = repeatedcv,
                              metric = "RMSE", 
                              tuneGrid = expand.grid(alpha = seq(0, 1, length = 20),
                                                     lambda = 10^seq(-1, 5, length = 20)))
  elastic_model_cpg = glmnet(x = makeX(train_x),
                           y = train_y,
                           alpha = elastic_tuning_cpg$bestTune[[1]], 
                           lambda = elastic_tuning_cpg$bestTune[[2]],
                           standardize = FALSE)
  return(elastic_model_cpg)
}
```

```{r, message = FALSE, warning = FALSE}
test_rmse = lapply(c(10, 25, 50, 100, 150, 200, 250), function(x) {select_top_n(x)}) %>%
  Reduce(rbind, .) %>%
  as.data.frame() %>%
  mutate(n = c(10, 25, 50, 100, 150, 200, 250))
```

```{r}
colnames(test_rmse) = c("lm", "age", "woage", "n")
test_rmse_melt = melt(test_rmse, id = "n") %>% as.data.frame()
pdf(file = "~/Desktop/DNA methylation/figures_0718/suppl_fig9.pdf", width = 3, height = 2)
ggplot(data = test_rmse_melt,
       aes(x = n, 
           y = value)) +
   geom_point(aes(color = as.factor(variable)), size = 1.25) + 
   geom_line(aes(group = as.factor(variable),
                 color = as.factor(variable))) +
   scale_color_manual(values = c("#117777", "#762A83", "#999933"))
```

```{r}
top_n_dmr = dmr_coeff_rank$dmr[1:100]
top_n_dmr_cpgs = fidmrs_cpg_combined %>% filter(dmr %in% top_n_dmr)
methy_dat_dis_cpg_associated = methy_dat_dis_merge %>%
    dplyr::select(logFI, sex_bin, age, top_n_dmr_cpgs$IlmnID)
methy_dat_val_cpg_associated = methy_dat_val_merge %>%
    dplyr::select(logFI, sex_bin, age, top_n_dmr_cpgs$IlmnID)
methy_dat_dis_cpg_associated_woage = methy_dat_dis_merge %>%
    dplyr::select(logFI, sex_bin, top_n_dmr_cpgs$IlmnID)
methy_dat_val_cpg_associated_woage = methy_dat_val_merge %>%
    dplyr::select(logFI, sex_bin, top_n_dmr_cpgs$IlmnID)
train_y_1 = methy_dat_dis_cpg_associated[, 1]
train_x_1 = methy_dat_dis_cpg_associated[, -1]
train_y_2 = methy_dat_dis_cpg_associated_woage[, 1]
train_x_2 = methy_dat_dis_cpg_associated_woage[, -1]
model1 = glmnet_cv(train_y_1, train_x_1, methy_dat_dis_cpg_associated)
model2 = glmnet_cv(train_y_2, train_x_2, methy_dat_dis_cpg_associated_woage)

frail_pred_test = methy_dat_val_merge %>% 
                    as.data.frame() %>%
                    mutate(lm_pred = unname(predict(methy_dat_FI_model, methy_dat_val_merge[, c("age", "sex_bin")])),
                           cpg_pred = predict(model1, as.matrix(methy_dat_val_cpg_associated[, -1]))[, 1],
                           cpg_age_pred = predict(model2, as.matrix(methy_dat_val_cpg_associated_woage[, -1]))[, 1]) 
frail_pred_train = methy_dat_dis_merge %>% 
                    as.data.frame() %>%
                    mutate(lm_pred = unname(predict(methy_dat_FI_model, methy_dat_dis_merge[, c("age", "sex_bin")])),
                           cpg_pred = predict(model1, as.matrix(methy_dat_dis_cpg_associated[, -1]))[, 1],
                           cpg_age_pred = predict(model2, as.matrix(methy_dat_dis_cpg_associated_woage[, -1]))[, 1]) 
```

```{r}
sessionInfo()
```




